### filter_ensemble_mutations.R ####################################################################
# FILTERs ensemble mutations (using OncoKB annotations as a white list) and plots results
# replaces output from plot_snv_summary.R for use in report

### FUNCTIONS ######################################################################################
# function to generate a standardized filename
generate.filename <- function(project.stem, file.core, extension, include.date = TRUE) {

	# build up the filename
	file.name <- paste(project.stem, file.core, sep = '_');
	file.name <- paste(file.name, extension, sep = '.');

	if (include.date) {
		file.name <- paste(Sys.Date(), file.name, sep = '_');
		}

	return(file.name);
	}

# function to write session profile to file
save.session.profile <- function(file.name) {

	# open the file
	sink(file = file.name, split = FALSE);

	# write memory usage to file
	cat('### MEMORY USAGE ###############################################################');
	print(proc.time());

	# write sessionInfo to file
	cat("\n### SESSION INFO ###############################################################");
	print(sessionInfo());

	# close the file
	sink();

	}

### PREPARE SESSION ################################################################################
# import libraries
library(xtable);
library(BoutrosLab.plotting.general);
library(argparse);

# import command line arguments
parser <- ArgumentParser();

parser$add_argument('-p', '--project', type = 'character', help = 'PROJECT name');
parser$add_argument('-o', '--output', type = 'character', help = 'path to output directory');
parser$add_argument('-t', '--seq_type', type = 'character', help = 'exome or wgs', default = 'exome');
parser$add_argument('-m', '--maf', type = 'character', help = 'mutation calls in MAF format');
parser$add_argument('-c', '--callable', type = 'character', help = 'path to callable bases file (generated by count_callable_bases.R)');

arguments <- parser$parse_args();

### VARIANT CODING
# 1 = missense, 2 = stop gain, 3 = stop loss, 4 = splicing, 5 = frameshift, 6 = in frame indel, 7 = tss
# 8 = RNA, 9 = other (up/downstream, UTR, intergenic, silent, intron), 10 = ITD
variant.codes <- data.frame(
	Classification = c("3'Flank", "5'Flank", "Intron", "RNA", "IGR", "3'UTR", "5'UTR", "Silent",
		"Missense_Mutation", "Splice_Region", "Splice_Site", "In_Frame_Del", "In_Frame_Ins",
		"Frame_Shift_Del", "Frame_Shift_Ins", "Nonsense_Mutation", "Nonstop_Mutation",
		"Translation_Start_Site", "ITD"),
	Group = c('other','other','other','RNA','other','other','other','other','missense',
		'splice_site','splice_site','in_frame_indel','in_frame_indel','frameshift_indel',
		'frameshirt_indel', 'nonsense', 'nonstop', 'tss', 'itd'),
	Code = c(9, 9, 9, 8, 9, 9, 9, 9, 1, 4, 4, 6, 6, 5, 5, 2, 3, 7, 10)
	);

variant.colours <- c('darkseagreen4','darkorchid4','#9AA3F2','yellow','darkorange3','#F9B38E','turquoise1','plum','grey50')
names(variant.colours) <- c('missense','nonsense','nonstop','splicing','frameshift_indel','in_frame_indel','tss','RNA','other');

# for these plots, we will ignore some variant types
variant.colours <- variant.colours[c(1:6,9)];
basechange.colours <- default.colours(8,'pastel')[-5];

# read in list of known driver genes (Cancer Gene Census from COSMIC)
driver.genes <- read.delim(
	'/cluster/projects/pughlab/references/COSMIC/cancer_gene_census_20210414.tsv',
	comment.char = '#'
	);

driver.genes <- driver.genes[which(driver.genes$Hallmark == 'Yes'),];

### READ DATA #####################################################################################
# get data
input.data <- read.delim(arguments$maf);

# stop if no OncoKB annotations available
if (! 'GENE_IN_ONCOKB' %in% colnames(input.data)) {
	stop('No OncoKB annotations available in input maf; please run OncoKB first.');
	}

callable.bases <- read.delim(arguments$callable, row.names = 1);

# collect list of all samples
all.samples <- sort(as.character(unique(input.data$Tumor_Sample_Barcode)));
alt.sample.names <- all.samples;

# ensure sample names match
if (!any(all.samples %in% rownames(callable.bases))) {
	if ( (any(grepl('-', all.samples))) & (!any(grepl('-', rownames(callable.bases)))) ) {
		alt.sample.names <- gsub('-', '.', all.samples);
		}
	if ( (any(grepl("^[[:digit:]]", all.samples))) & (any(grepl('^X',rownames(callable.bases)))) ) {
		alt.sample.names <- paste0('X', alt.sample.names);
		}
	if (!any(alt.sample.names %in% rownames(callable.bases))) {
		stop('Correlation sample names do not match names from other tables');
		}
	callable.bases <- callable.bases[alt.sample.names,];
	callable.bases$Sample <- rownames(callable.bases);
	rownames(callable.bases) <- all.samples;
	} else {
	callable.bases <- callable.bases[all.samples,];
	}

# move to output directory
setwd(arguments$output);

### FILTER DATA ####################################################################################
# oncokb seems to select genes *purely* on gene symbol (including alternate names)
# for example, TEP1 is selected because of PTEN (aka TEP1), even though they are
# on different chromosomes
oncokb.columns <- c(grep('GENE_IN_ONCOKB', colnames(input.data)):ncol(input.data));
known.bugs <- which(input.data$Hugo_Symbol %in% c('PRMT9','RAD1','TEP1','FAH'));
if (length(known.bugs) > 0) {
	input.data[known.bugs,oncokb.columns] <- NA;
	}

# find probable false positives to remove
remove.these <- which(input.data$FLAG.tumour_only & input.data$FLAG.high_pop);
also.remove.these <- which(
	input.data$FLAG.tumour_only & 
	input.data$FLAG.low_vaf
	);

# but whitelist these
keep.these <- grep('Oncogenic', input.data$ONCOGENIC);

# apply filter
if (length(setdiff(unique(c(remove.these,also.remove.these)), keep.these)) > 0) {
	input.data <- input.data[-setdiff(unique(c(remove.these,also.remove.these)), keep.these),];
	}

# extract key hits
significant.hits <- which(
	input.data$ONCOGENIC == 'Oncogenic' |
	input.data$MUTATION_EFFECT %in% c('Loss-of-function','Likely Loss-of-function') |
	input.data$MUTATION_EFFECT %in% c('Switch-of-function') |
	input.data$MUTATION_EFFECT %in% c('Gain-of-function','Likely Gain-of-function')
	);

# format for latex
oncokb.hits <- input.data[significant.hits,];

oncokb.to.print <- data.frame(
	ID = oncokb.hits$Tumor_Sample_Barcode,
	Gene = oncokb.hits$Hugo_Symbol,
	Position = oncokb.hits$Start_Position,
	VAF = round(oncokb.hits$t_alt_count / oncokb.hits$t_depth,2),
	VARIANT_IN_ONCOKB = as.logical(oncokb.hits$VARIANT_IN_ONCOKB),
	Effect = oncokb.hits$MUTATION_EFFECT
	);

# order by count and median VAF
gene.counts <- aggregate(VAF ~ Gene, oncokb.to.print, length);
colnames(gene.counts)[2] <- 'Count';
gene.counts$VAF <- aggregate(VAF ~ Gene, oncokb.to.print, median)$VAF;
gene.counts <- gene.counts[order(-gene.counts$Count, -gene.counts$VAF),];
gene.counts <- gene.counts[1:min(40,nrow(gene.counts)),];

oncokb.to.print$Gene <- factor(oncokb.to.print$Gene, levels = gene.counts$Gene);
oncokb.to.print <- oncokb.to.print[!is.na(oncokb.to.print$Gene),];
oncokb.to.print <- oncokb.to.print[order(
	oncokb.to.print$Gene,
	oncokb.to.print$VAF
	),];

### FORMAT DATA ####################################################################################
# indicate key fields
keep.fields <- c('Tumor_Sample_Barcode','Hugo_Symbol','Chromosome','Variant_Classification','Reference_Allele','Tumor_Seq_Allele2','t_depth','t_alt_count','VARIANT_IN_ONCOKB','ONCOGENIC','MUTATION_EFFECT');
mutation.data <- input.data[,keep.fields];

# apply variant coding
mutation.data$Code <- variant.codes$Code[match(mutation.data$Variant_Classification, variant.codes$Classification)];
mutation.data[which(mutation.data$Code > 6),]$Code <- 9;

# get per-sample mutation counts
sample.counts <- data.frame(cbind(
	table(mutation.data$Tumor_Sample_Barcode),
	table(mutation.data[which(mutation.data$Code != 9),]$Tumor_Sample_Barcode)
	));
colnames(sample.counts) <- c('Total','Non.Silent');

# convert these to mutation rates
sample.counts$Callable <- NA;

for (smp in rownames(sample.counts)) {
	if (arguments$seq_type == 'wgs') {
		sample.counts[smp,]$Callable <- callable.bases[smp,]$Patient.total;
		} else {
		sample.counts[smp,]$Callable <- callable.bases[smp,]$Patient.targeted;
		}
	}

sample.counts$SNVsPerMb <- sample.counts$Total / sample.counts$Callable * 10**6;
sample.counts$NonSilent.Rate.Mb <- sample.counts$Non.Silent / sample.counts$Callable * 10**6;
sample.counts <- sample.counts[order(sample.counts$SNVsPerMb, decreasing = TRUE),];
sample.counts$Order <- 1:nrow(sample.counts);

# summarize mutations
functional.summary <- data.frame(table(mutation.data[,c('Tumor_Sample_Barcode','Code')]));
tmp <- aggregate(Freq ~ Tumor_Sample_Barcode,functional.summary,sum);
functional.summary <- merge(functional.summary, tmp, by = 'Tumor_Sample_Barcode');
functional.summary$Proportion <- functional.summary$Freq.x / functional.summary$Freq.y;
functional.summary$Tumor_Sample_Barcode <- factor(
	functional.summary$Tumor_Sample_Barcode,
	levels = rownames(sample.counts)
	);
functional.summary$Code <- factor(functional.summary$Code, levels = rev(c(1:6,9)));
rm(tmp);

alt.functional <- data.frame(table(mutation.data[which(mutation.data$Code < 9),c('Tumor_Sample_Barcode','Code')]));
tmp <- aggregate(Freq ~ Tumor_Sample_Barcode, alt.functional, sum);
alt.functional <- merge(alt.functional, tmp, by = 'Tumor_Sample_Barcode');
alt.functional$Proportion <- alt.functional$Freq.x / alt.functional$Freq.y;
alt.functional$Tumor_Sample_Barcode <- factor(
	alt.functional$Tumor_Sample_Barcode,
	levels = rownames(sample.counts)
	);
alt.functional$Code <- factor(alt.functional$Code, levels = rev(c(1:6,9)));
rm(tmp);

mutation.data$Basechange <- paste0(mutation.data$Reference_Allele, '>', mutation.data$Tumor_Seq_Allele2);
mutation.data[which(input.data$Variant_Type %in% c('DEL', 'INS')),]$Basechange <- 'indel';
mutation.data[which(mutation.data$Basechange == 'T>G'),]$Basechange <- 'A>C';
mutation.data[which(mutation.data$Basechange == 'T>C'),]$Basechange <- 'A>G';
mutation.data[which(mutation.data$Basechange == 'T>A'),]$Basechange <- 'A>T';
mutation.data[which(mutation.data$Basechange == 'C>T'),]$Basechange <- 'G>A';
mutation.data[which(mutation.data$Basechange == 'C>G'),]$Basechange <- 'G>C';
mutation.data[which(mutation.data$Basechange == 'C>A'),]$Basechange <- 'G>T';
mutation.data$Basechange <- factor(
	mutation.data$Basechange,
	levels = rev(c('A>C','A>G','A>T','G>A','G>C','G>T','indel'))
	);
basechange.summary <- data.frame(table(mutation.data[,c('Tumor_Sample_Barcode','Basechange')]));
tmp <- aggregate(Freq ~ Tumor_Sample_Barcode, basechange.summary, sum);
basechange.summary <- merge(basechange.summary, tmp, by = 'Tumor_Sample_Barcode');
basechange.summary$Proportion <- basechange.summary$Freq.x / basechange.summary$Freq.y;
basechange.summary$Tumor_Sample_Barcode <- factor(
	basechange.summary$Tumor_Sample_Barcode,
	levels = rownames(sample.counts)
	);
rm(tmp);

# remove silent mutations
remove.idx <- which(mutation.data$Variant_Classification %in% c('Silent','Intron','IGR'));
mutation.data <- mutation.data[-remove.idx,];

# reduce to 1 mutation per gene per sample [taking the higher priority code]
mutation.data.trimmed <- aggregate(
	Code ~ Tumor_Sample_Barcode + Hugo_Symbol + Chromosome,
	mutation.data,
	min
	);

# reshape data
plot.data <- reshape(
	mutation.data.trimmed,
	direction = 'wide',
	timevar = 'Tumor_Sample_Barcode',
	idvar = c('Chromosome','Hugo_Symbol')
	);
colnames(plot.data) <- gsub('Code.','',colnames(plot.data));

# ensure all samples are accounted for
absent.smps <- setdiff(all.samples, colnames(plot.data));
if (length(absent.smps) > 0) { plot.data[,absent.smps] <- NA; }

# get per-gene sample counts
plot.data <- plot.data[,c('Hugo_Symbol', 'Chromosome', all.samples)];
plot.data$Count <- apply(plot.data[,all.samples],1,function(i) { length(i[which(i < 9)]) } );
plot.data$Priority <- apply(plot.data[,all.samples],1,function(i) { sum(i[which(i < 9)]) } )/plot.data$Count;
plot.data <- plot.data[order(-plot.data$Count, plot.data$Priority),];

# order samples by recurrence for prettier heatmap
heatmap.data <- t(plot.data[,all.samples]);
heatmap.data[!is.na(heatmap.data)] <- 1;
heatmap.data <- heatmap.data[do.call(order, transform(heatmap.data)),];

# remove non-recurrently mutated genes
plot.data <- plot.data[which(plot.data$Count > 1),];

recurrence.summary <- as.data.frame(table(plot.data[which(plot.data$Count > 1),]$Count));
colnames(recurrence.summary) <- c('N.samples','N.genes');

# remove known problem (passenger) genes
genes.to.exclude <- c('TTN','MUC17','OBSCN','MUC16','SYNE2','NEB','SYNE1');
plot.data <- plot.data[which(!plot.data$Hugo_Symbol %in% genes.to.exclude),];

# trim down to top recurrently mutated genes (or top 20, whichever is smallest)
tophit.flag <- FALSE;
top.count <- which(plot.data$Count > length(all.samples)*0.2);
if (length(top.count) <= 20 & length(top.count) > 2) {
	plot.data <- plot.data[which(plot.data$Count > length(all.samples)*0.2),];
	} else if (nrow(plot.data[which(plot.data$Count == length(all.samples)),]) > 20) {
	# likely only true for datasets with no matched normal
	# plot known tumour suppressors
	tophit.flag <- TRUE;
	plot.data <- plot.data[which(plot.data$Hugo_Symbol %in% driver.genes$Hugo_Symbol),];
	}

if (nrow(plot.data) > 20) {
	plot.data <- plot.data[1:20,];
	}

# clean up for plotting
plot.data$Label <- as.character(plot.data$Hugo_Symbol);
if (any(duplicated(plot.data$Hugo_Symbol))) {
	dup.symbols <- unique(plot.data$Hugo_Symbol[duplicated(plot.data$Hugo_Symbol)]);
	for (gene in dup.symbols) {
		tmp <- plot.data[which(plot.data$Hugo_Symbol == gene),1:2];
		new.symbols <- paste0(plot.data$Hugo_Symbol, '_', plot.data$Chromosome);
		plot.data[which(plot.data$Hugo_Symbol == gene),]$Label <- new.symbols;
		}
	}

### MAKE PLOTS #####################################################################################
# make the plot legend (mutation type/consequence)
functional.legend <- legend.grob(
	legends = list(
		legend = list(
			colours = variant.colours,
			labels = names(variant.colours)
			)
		),
	title.just = 'left',
	size = 2
	);

basechange.legend <- legend.grob(
	legends = list(
		legend = list(
			colours = basechange.colours,
			labels = c(
				expression('A' %->% 'C,T' %->% 'G'),
				expression('A' %->% 'G,T' %->% 'C'),
				expression('A' %->% 'T,T' %->% 'A'),
				expression('G' %->% 'A,C' %->% 'T'),
				expression('G' %->% 'C,C' %->% 'G'),
				expression('G' %->% 'T,C' %->% 'A'),
				'indel'
				)
			)
		),
	title.just = 'left',
	size = 2
	);

# grab some parameters
axis.cex <- if (length(all.samples) <= 30) { 1
	} else if (length(all.samples) <= 50) { 0.75
	} else if (length(all.samples) <= 80) { 0.5
	} else { 0 };

# create heatmap for recurrent genes (ordered by recurrence)
create.heatmap(
	plot.data[,rownames(heatmap.data)],
	cluster.dimensions = 'none',
	same.as.matrix = TRUE,
	xaxis.lab = rownames(heatmap.data),
	yaxis.lab = plot.data$Label,
	xaxis.cex = axis.cex,
	yaxis.cex = 1,
	xaxis.tck = if (axis.cex == 0) { 0 } else { 0.2 },
	yaxis.tck = 0.2,
	xaxis.fontface = 'plain',
	yaxis.fontface = 'plain',
	axes.lwd = 1,
	grid.row = TRUE,
	force.grid.row = TRUE,
	row.colour = 'grey80',
	col.colour = 'grey80',
	row.lwd = if (length(all.samples) < 30) { 3 } else { 1 },
	col.lwd = if (length(all.samples) < 30) { 3 } else { 1 },
	grid.col = TRUE,
	force.grid.col = TRUE,
	print.colour.key = FALSE,
	fill.colour = 'white',
	at = seq(0,length(variant.colours),1),
	total.colours = length(variant.colours)+1,
	colour.scheme = variant.colours,
	inside.legend = if (length(all.samples) < 12) { 
		list(fun = functional.legend, x = 1.02, y = 1)
		} else { NULL },
	right.padding = if (length(all.samples) < 12) { 21 } else { 0 },
	height = 6,
	width = 8,
	resolution = 200,
	filename = generate.filename(arguments$project, 'snv_recurrent_genes_filtered','png')
	);

# create plot for mutation rates
ylim <- c(floor(min(log10(sample.counts$SNVsPerMb))), ceiling(max(log10(sample.counts$SNVsPerMb))));
yat <- seq(ylim[1], ylim[2], 1);
yaxis.labels <- 10**yat;
sample.counts$Rate <- log10(sample.counts$SNVsPerMb);

if (ylim[2] == 1) {
	ylim <- c(0, ceiling(max(sample.counts$SNVsPerMb)));
	yat <- if (ylim[2] > 5) { seq(0,ylim[2],2); } else { seq(0,ylim[2],1); }
	yaxis.labels <- yat;
	sample.counts$Rate <- sample.counts$SNVsPerMb;
	}

rate.plot <- create.scatterplot(
	Rate ~ Order,
	sample.counts,
	xaxis.lab = rep('', nrow(sample.counts)+1),
	yaxis.cex = 1,
	xaxis.tck = 0,
	yaxis.tck = c(0.5,0),
	yaxis.fontface = 'plain',
	axes.lwd = 1,
	ylab.label = 'SNVs/Mbp',
	ylab.cex = 1.2,
	ylimits = ylim,
	yat = yat,
	yaxis.lab = yaxis.labels,
	xlab.label = NULL,
	xlimits = c(0.5,nrow(sample.counts)+0.5)
	);

# create plot for functional summary
mutation.type.plot <- create.barplot(
	Proportion ~ Tumor_Sample_Barcode,
	functional.summary,
	groups = functional.summary$Code,
	stack = TRUE,
	col = rev(variant.colours),
	xaxis.lab = rep('',length(all.samples)),
	xlimits = c(0.5, length(all.samples)+0.5),
	xat = seq(1,length(all.samples)),
	yaxis.tck = c(0.5,0),
	xaxis.tck = 0,
	xaxis.fontface = 'plain',
	yaxis.fontface = 'plain',
	ylab.label = 'Proportion',
	ylab.cex = 1.2,
	xlab.label = NULL,
	ylimits = c(0,1),
	yat = seq(0,1,0.5),
	yaxis.cex = 1,
	axes.lwd = 1,
	top.padding = 0
	);

functional.plot <- create.barplot(
	Proportion ~ Tumor_Sample_Barcode,
	alt.functional,
	groups = alt.functional$Code,
	stack = TRUE,
	col = rev(variant.colours),
	xaxis.lab = rep('',length(all.samples)),
	xlimits = c(0.5, length(all.samples)+0.5),
	xat = seq(1,length(all.samples)),
	yaxis.tck = c(0.5,0),
	xaxis.tck = 0,
	xaxis.fontface = 'plain',
	yaxis.fontface = 'plain',
	ylab.label = 'Proportion',
	ylab.cex = 1.2,
	xlab.label = NULL,
	ylimits = c(0,1),
	yat = seq(0,1,0.5),
	yaxis.cex = 1,
	axes.lwd = 1,
	top.padding = 0
	);

# create plot for basechange summary
basechange.plot <- create.barplot(
	Proportion ~ Tumor_Sample_Barcode,
	basechange.summary,
	groups = basechange.summary$Basechange,
	stack = TRUE,
	col = rev(basechange.colours),
	xaxis.lab = levels(basechange.summary$Tumor_Sample_Barcode),
	xaxis.rot = 90,
	xlimits = c(0.5, length(all.samples)+0.5),
	xat = seq(1,length(all.samples)),
	yaxis.tck = c(0.5,0),
	xaxis.tck = 0,
	xaxis.fontface = 'plain',
	yaxis.fontface = 'plain',
	ylab.label = 'Proportion',
	ylab.cex = 1.2,
	xlab.label = NULL,
	ylimits = c(0,1),
	yat = seq(0,1,0.5),
	yaxis.cex = 1,
	xaxis.cex = axis.cex,
	axes.lwd = 1,
	top.padding = 0,
	legend = list(
		right = list(fun = basechange.legend)
		)
	);

# combine them!
if (arguments$seq_type == 'wgs') {

	create.multipanelplot(
		plot.objects = list(rate.plot, functional.plot, basechange.plot),
		height = 10,
		width = 8,
		resolution = 200,
		filename = generate.filename(arguments$project, 'mutation_summary_filtered','png'),
		plot.objects.heights = c(2,3,4),
		left.legend.padding = 0,
		right.legend.padding = 0,
		top.legend.padding = 0,
		bottom.legend.padding = 0,
		right.padding = 2,
		y.spacing = -1,
		ylab.axis.padding = 0.5,
		legend = list(
			inside = list(fun = functional.legend, x = 0.88, y = 0.65)
			)
		);

	} else {
	create.multipanelplot(
		plot.objects = list(rate.plot, mutation.type.plot, functional.plot, basechange.plot),
		height = 10,
		width = 8,
		resolution = 200,
		filename = generate.filename(arguments$project, 'mutation_summary_filtered','png'),
		plot.objects.heights = c(1.5,2,2,4),
		left.legend.padding = 0,
		right.legend.padding = 0,
		top.legend.padding = 0,
		bottom.legend.padding = 0,
		right.padding = 2,
		y.spacing = -1,
		ylab.axis.padding = 0.5,
		legend = list(
			inside = list(fun = functional.legend, x = 0.88, y = 0.65)
			)
		);
	}

# plot oncoKB hits (if present)
effect.colours <- c(default.colours(8,'div')[c(2,4,6,8)],'purple4');
names(effect.colours) <- c('Loss-of-function','Likely Loss-of-function','Likely Gain-of-function','Gain-of-function','Switch-of-function');

create.boxplot(
	VAF ~ Gene,
	oncokb.to.print,
	add.stripplot = TRUE,
	points.col = effect.colours[match(oncokb.to.print$Effect, names(effect.colours))],
	points.alpha = 0.8,
	points.cex = 0.8,
	xaxis.rot = 90,
	yaxis.tck = c(0.5,0),
	xaxis.tck = c(0.5,0),
	ylab.label = 'Variant Allele Frequency',
	ylab.cex = 1.2,
	ylab.axis.padding = 1,
	xlab.label = NULL,
	ylimits = c(0,1),
	yat = seq(0,1,0.2),
	yaxis.cex = 1,
	xaxis.cex = if (nrow(gene.counts) < 30) { 1 } else { 0.75 },
	style = 'Nature',
	key = list(
		points = list(col = effect.colours, pch = 19, cex = 1),
		text = list(lab = names(effect.colours), cex = 0.8, col = 'black'),
		title = 'Mutation Effect',
		cex.title = 1.2,
		x = if (nrow(gene.counts) >= 30) { 0.7 } else { 0.6 },
		y = 1
		),
	filename = generate.filename(arguments$project, 'oncoKB_hits', 'png'),
	height = 4,
	width = if (nrow(gene.counts) < 30) { 6 } else { 8 }
	);

save(
	sample.counts,
	functional.summary,
	basechange.summary,
	plot.data,
	variant.colours,
	oncokb.hits,
	file = generate.filename(arguments$project, 'mutation_summary_filtered', 'RData')
	);

### LATEX ##########################################################################################
# write some captions
if (arguments$seq_type == 'wgs') {
	summary.caption <- "Summary of short somatic variants (SNVs, indels) for each sample. From top to bottom: mutation rate (SNVs and INDELs per Mbp); proportion of all non-silent mutations with indicated functional consequence; proportion of total mutations with the indicated mutation type.";
	} else {
	summary.caption <- "Summary of short somatic variants (SNVs, indels) for each sample. From top to bottom: mutation rate (SNVs and INDELs per Mbp); proportion of total mutations with indicated functional consequence; proportion of all non-silent mutations with indicated functional consequence; proportion of total mutations with the indicated mutation type.";
	}

if (tophit.flag) {
	recurrence.caption <- "Summary of the top 20 most frequently mutated known driver genes across the cohort. Colours indicate predicted functional consequence (white = no mutation detected).";
	} else {
	recurrence.caption <- "Summary of the most frequently mutated genes across the cohort. Figure shows genes with mutations detected in $>$20\\% of samples (or top 20 genes). Colours indicate predicted functional consequence (white = no mutation detected).";
	}

# write for latex
write("\\section{SNV Summary}", file = 'somatic_snv_summary_filtered.tex');

# first, check for mutation_overlap plot
overlap.plot <- rev(sort(list.files(pattern = 'mutation_overlap.png')));
if (length(overlap.plot) > 0) {
	write("\\begin{figure}[h!]", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	write("\\begin{center}", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	write(paste0(
		"\\includegraphics[width=0.8\\textwidth]{",
		getwd(), '/',
		overlap.plot[1], '}'
		), file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	write("\\end{center}", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	write(paste0(
		"\\caption{From top to bottom: number of variants called by each tool (Mutect, MuTect2, SomaticSniper, Strelka, VarDict and VarScan) and the ensemble variants carried forward.}"
		), file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	write("\\end{figure}\n\\pagebreak\n", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	}

# describe filtering (if any tumour-only)
if (any(input.data$FLAG.tumour_only)) {
	write('Ensemble calls were filtered to remove probable false positives from tumour-only samples using the following criteria:',	file = 'somatic_snv_summary_filtered.tex', append = TRUE);

	write("\\begin{itemize}", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	write("\\item remove any variant with AF $>$ 0.001 in any population", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	write("\\item remove any variant with VAF $<$ 0.05", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	write("\\item keep any variant identified by OncoKB as oncogenic or ", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	write("\\item keep any variant identified by OncoKB as Loss-of-function, Likely Loss-of-function, Gain-of-function, Likely Gain-of-function or Switch-of-function", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	write("\\end{itemize}", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	}

# summarize key OncoKB findings
if (nrow(oncokb.hits) > 0) {
	write("\\vskip 1cm\n\\textbf{OncoKB}\n", file = 'somatic_snv_summary_filtered.tex', append = TRUE);

	oncokb.caption <- paste0('OncoKB was used to annotate known/suspected oncogenic variants; this identified ', nlevels(droplevels(oncokb.hits$Hugo_Symbol)), ' genes (', nrow(oncokb.hits), ' total mutations) across the cohort.');

	if (nrow(gene.counts) > 5) {
		write("\\begin{figure}[h!]", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
		write("\\begin{center}", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
		write(paste0(
			"\\includegraphics[width=0.9\\textwidth]{",
			getwd(), '/',
			generate.filename(arguments$project, 'oncoKB_hits', 'png'), '}'
			), file = 'somatic_snv_summary_filtered.tex', append = TRUE);
		write("\\end{center}", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
		write(paste0(
			"\\caption{", paste0(oncokb.caption, " Top hits are shown, sorted by recurrence and median VAF."), "}"
			), file = 'somatic_snv_summary_filtered.tex', append = TRUE);
		write("\\end{figure}\n", file = 'somatic_snv_summary_filtered.tex', append = TRUE);

		write("\\pagebreak\n", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
		}

	if (nrow(oncokb.to.print) <= 20) {
		print(
			xtable(
				oncokb.to.print,
				caption = oncokb.caption
				),
			file = 'somatic_snv_summary_filtered.tex',
			include.rownames = FALSE,
			size = 'scriptsize',
			append = TRUE
			);
		}

	write("\\pagebreak\n", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	}

# summarize filtered dataset	
write("\\textbf{Filtered Mutation Summary}\n", file = 'somatic_snv_summary_filtered.tex', append = TRUE);

write("\\begin{figure}[h!]", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
write("\\begin{center}", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
write(paste0(
	"\\includegraphics[width=0.8\\textwidth]{",
	getwd(), '/',
	generate.filename(arguments$project, 'mutation_summary_filtered','png'), '}'
	), file = 'somatic_snv_summary_filtered.tex', append = TRUE);
write("\\end{center}", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
write(paste0(
	"\\caption{", summary.caption, "}"
	), file = 'somatic_snv_summary_filtered.tex', append = TRUE);
write("\\end{figure}\n", file = 'somatic_snv_summary_filtered.tex', append = TRUE);

# summarize recurrent mutations
if (nrow(plot.data) == 0) {
	write("No genes were recurrently mutated across the cohort.", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	} else {
	write("\\pagebreak\n", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	print(
		xtable(
			rev(recurrence.summary)[1:(min(nrow(recurrence.summary),10)),],
			caption = 'Summary of recurrently-mutated genes across the cohort. Table shows number of genes with functionally relevant mutations in N samples.'
			),
		file = 'somatic_snv_summary_filtered.tex',
		include.rownames = FALSE,
		append = TRUE
		);

	write("\n\\begin{figure}[h!]", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	write("\\begin{center}", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	write(paste0(
		"\\includegraphics[width=0.8\\textwidth]{",
		getwd(), '/',
		generate.filename(arguments$project, 'snv_recurrent_genes_filtered','png'), '}'
		), file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	write("\\end{center}", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	write(paste0(
		"\\caption{", recurrence.caption, "}"
		), file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	write("\\end{figure}\n", file = 'somatic_snv_summary_filtered.tex', append = TRUE);
	}

### SAVE SESSION INFO ##############################################################################
save.session.profile(generate.filename('MutationSummaryFiltered','SessionProfile','txt'));
