### plot_sv_summary.R ############################################################################
# Identify and plot recurrent SVs.
# INPUT:
#	- sv calls (combined mavis output)

### FUNCTIONS ######################################################################################
# function to generate a standardized filename
generate.filename <- function(project.stem, file.core, extension, include.date = TRUE) {

	# build up the filename
	file.name <- paste(project.stem, file.core, sep = '_');
	file.name <- paste(file.name, extension, sep = '.');

	if (include.date) {
		file.name <- paste(Sys.Date(), file.name, sep = '_');
		}

	return(file.name);
	}

# function to write session profile to file
save.session.profile <- function(file.name) {

	# open the file
	sink(file = file.name, split = FALSE);

	# write memory usage to file
	cat('### MEMORY USAGE ###############################################################');
	print(proc.time());

	# write sessionInfo to file
	cat("\n### SESSION INFO ###############################################################");
	print(sessionInfo());

	# close the file
	sink();

	}

### PREPARE SESSION ################################################################################
# import libraries
library(xtable);
library(BoutrosLab.plotting.general);
library(UpSetR);
library(argparse);

# import command line arguments
parser <- ArgumentParser();

parser$add_argument('-p', '--project', type = 'character', help = 'PROJECT name');
parser$add_argument('-o', '--output', type = 'character', help = 'path to output directory');
parser$add_argument('-m', '--mavis', type = 'character', help = 'combined variant calls (generated by collect_mavis_output.R)');

arguments <- parser$parse_args();

### READ DATA ######################################################################################
# get data
input.data <- read.delim(arguments$mavis);

# collect list of all samples
tumour.samples <- colnames(input.data)[grep('diseased_genome', colnames(input.data))];
tumour.samples <- as.character(sapply(tumour.samples, function(i) { unlist(strsplit(i,'_'))[1] } ));
tumour.samples <- gsub('^X', '', gsub('\\.', '-', tumour.samples));

normal.samples <- colnames(input.data)[grep('normal_genome', colnames(input.data))];
normal.samples <- as.character(sapply(normal.samples, function(i) { unlist(strsplit(i,'_'))[1] } ));
normal.samples <- gsub('^X', '', gsub('\\.', '-', normal.samples));

# trim down input data
sv.data <- input.data[,!grepl('genome', colnames(input.data))];
sv.data$library <- gsub('^X', '', gsub('\\.', '-', sv.data$library));

# move to output directory
setwd(arguments$output);

### FORMAT DATA ####################################################################################
# summarize tool data
tool.list <- c('Manta','Delly','Pindel','SViCT','NovoBreak'); 
tool.summary <- data.frame(
	Tool = tool.list,
	Total = c(
		nrow(sv.data[grepl('manta',sv.data$tool),]),
		nrow(sv.data[grepl('delly',sv.data$tool),]),
		nrow(sv.data[grepl('pindel',sv.data$tool),]),
		nrow(sv.data[grepl('svict',sv.data$tool),]),
		nrow(sv.data[grepl('novobreak',sv.data$tool),])
		)
	);

for (type in unique(sv.data$event_type)) {
	tmp <- sv.data[which(sv.data$event_type == type),];
	tool.summary[,type] <- c(
		nrow(tmp[grepl('manta',tmp$tool),]),
		nrow(tmp[grepl('delly',tmp$tool),]),
		nrow(tmp[grepl('pindel',tmp$tool),]),
		nrow(tmp[grepl('svict',tmp$tool),]),
		nrow(tmp[grepl('novobreak',tmp$tool),])
		);
	}

rownames(tool.summary) <- tool.summary$Tool;
tool.summary <- t(tool.summary[,-1]);

# now format data for plotting
tool.list <- unique(unlist(strsplit(as.character(sv.data$tools),';')));
tool.data <- as.data.frame(matrix(nrow = nrow(sv.data), ncol = length(tool.list), data = 0));
colnames(tool.data) <- tool.list;

for (tool in tool.list) {
	idx <- grepl(tool, sv.data$tools);
	tool.data[idx,tool] <- 1;
	}

# count variants
sample.counts <- data.frame(table(sv.data$library));
colnames(sample.counts) <- c('Sample','Count');
sample.counts <- sample.counts[order(sample.counts$Count),];

# summarize mutations
functional.summary <- data.frame(table(sv.data[,c('library','event_type')]));
functional.summary <- merge(functional.summary, sample.counts, by.x = 'library', by.y = 'Sample');
functional.summary$Proportion <- functional.summary$Freq / functional.summary$Count;
functional.summary$Tumor_Sample_Barcode <- factor(
	functional.summary$library,
	levels = sample.counts$Sample
	);

functional.summary$event_type <- factor(
	functional.summary$event_type,
	levels = c('deletion','duplication','insertion','inversion','inverted translocation','translocation')
	);

functional.summary <- functional.summary[order(functional.summary$Tumor_Sample_Barcode, functional.summary$event_type),];

# search for recurrent events
tmp <- sv.data[!grepl('None', sv.data$gene1_aliases) & !grepl('None', sv.data$gene2_aliases),];
tmp$Fusion <- paste0(tmp$gene1_aliases, '--', tmp$gene2_aliases);

# first, bin by breakpoint (1kbp)
tmp$start_chr <- tmp$break1_chromosome;
tmp$end_chr <- tmp$break2_chromosome;
tmp$start_bin <- as.integer(floor(tmp$break1_position_start/10**3)*10**3);
tmp$end_bin <- as.integer(ceiling(tmp$break2_position_end/10**3)*10**3);

event.data <- unique(
	tmp[,c('library','event_type','Fusion','start_chr','end_chr','start_bin','end_bin')]
	);

event.data$Event <- paste0(
	'chr', event.data$start_chr, ':', event.data$start_bin,
	'_chr', event.data$end_chr, ':', event.data$end_bin
	);

event.counts <- aggregate(
	library ~ Fusion + event_type + Event,
	event.data[event.data$library %in% tumour.samples,],
	length
	);
colnames(event.counts)[4] <- 'Count';

event.counts <- event.counts[order(event.counts$Count, decreasing = TRUE),];

# reshape data (for future plotting?)
#plot.data <- reshape(
#	event.counts,
#	direction = 'wide',
#	timevar = 'event_type',
#	idvar = c('Event','Fusion')
#	);
#colnames(plot.data) <- gsub('Count.','',colnames(plot.data));
#colnames(plot.data) <- gsub(' ', '_', colnames(plot.data));

event.counts <- event.counts[which(event.counts$Count > 1),c('Event','Fusion','event_type','Count')];

recurrence.data <- list();
for (type in unique(sv.data$event_type)) {
	tmp <- event.counts[which(event.counts$event_type == type),];
	if (nrow(tmp) == 0) { next; }
	if (nrow(tmp) > 10) { tmp <- tmp[1:10,]; }
	recurrence.data[[type]] <- tmp[,c(1,2,4)];
	}

### PLOT DATA ######################################################################################
# plot tool summary (overlap plot)
png(filename = generate.filename(arguments$project, 'SV_tool_overlap','png'),
	res = 200, units = 'in', height = 5, width = 7);

upset(tool.data, sets = tool.list, mainbar.y.label = 'SVs Called By Indicated ToolSet',
	sets.x.label = 'SVs Called Per Tool');

dev.off();

# now set up variables for SV plots
sv.colour.scheme <- rev(default.colours(12))[1:6]; #length(unique(sv.data$event_type))];
names(sv.colour.scheme) <- c('deletion','duplication','insertion','inversion','inverted translocation','translocation');

# make the plot legend (mutation type/consequence)
functional.legend <- legend.grob(
	legends = list(
		legend = list(
			colours = sv.colour.scheme[1:3],
			labels = names(sv.colour.scheme)[1:3]
			),
		legend = list(
			colours = sv.colour.scheme[4:6],
			labels = names(sv.colour.scheme)[4:6]
			)
		),
	layout = c(2,1),
	size = 2
	);

# grab some parameters
axis.cex <- if (length(normal.samples) <= 30) { 1
	} else if (length(normal.samples) <= 50) { 0.75
	} else if (length(normal.samples) <= 80) { 0.5
	} else { 0 };

plot.height <- if (length(normal.samples) <= 30) { 6
	} else if (length(normal.samples) <= 50) { 8
	} else { 9 }

legend.pos <- if (plot.height == 6) { 1.2
	} else { 1.1 }

# create plot for functional summary
if (length(normal.samples) > 1) {
	create.barplot(
		Tumor_Sample_Barcode ~ Freq,
		functional.summary[which(functional.summary$library %in% normal.samples),],
		groups = functional.summary$event_type,
		stack = TRUE,
		plot.horizontal = TRUE,
		col = sv.colour.scheme,
		yaxis.tck = if (axis.cex == 0) { 0 } else { c(0.5,0) },
		xaxis.tck = c(0.5,0),
		xaxis.fontface = 'plain',
		yaxis.fontface = 'plain',
		xlab.label = 'Count',
		xlab.cex = 1.2,
		ylab.label = NULL,
		yaxis.cex = axis.cex,
		xaxis.cex = 1,
		axes.lwd = 1,
		top.padding = 10,
		legend = list(
			inside = list(fun = functional.legend, x = 0.1, y = legend.pos)
			),
		height = plot.height,
		width = 7,
		resolution = 200,
		filename = generate.filename(arguments$project, 'germline_sv_summary', 'png')
		);
	}

# grab some parameters
axis.cex <- if (length(tumour.samples) <= 30) { 1
	} else if (length(tumour.samples) <= 50) { 0.75
	} else if (length(tumour.samples) <= 80) { 0.5
	} else { 0 };

plot.height <- if (length(tumour.samples) <= 30) { 6
	} else if (length(tumour.samples) <= 50) { 8
	} else { 9 }

legend.pos <- if (plot.height == 6) { 1.2
	} else { 1.1 }

# create plot for functional summary
create.barplot(
	Tumor_Sample_Barcode ~ Freq,
	functional.summary[which(functional.summary$library %in% tumour.samples),],
	groups = functional.summary$event_type,
	stack = TRUE,
	plot.horizontal = TRUE,
	col = sv.colour.scheme,
	yaxis.tck = if (axis.cex == 0) { 0 } else { c(0.5,0) },
	xaxis.tck = c(0.5,0),
	xaxis.fontface = 'plain',
	yaxis.fontface = 'plain',
	xlab.label = 'Count',
	xlab.cex = 1.2,
	ylab.label = NULL,
	yaxis.cex = axis.cex,
	xaxis.cex = 1,
	axes.lwd = 1,
	top.padding = 10,
	legend = list(
		inside = list(fun = functional.legend, x = 0.1, y = legend.pos)
		),
	height = plot.height,
	width = 7,
	resolution = 200,
	filename = generate.filename(arguments$project, 'somatic_sv_summary', 'png')
	);

save(
	sv.data,
	tool.summary,
	functional.summary,
	event.data,
	sv.colour.scheme,
	file = generate.filename(arguments$project, 'sv_summary', 'RData')
	);

# write for latex
write("\\section{SV Summary}", file = 'sv_summary.tex');

if (nrow(sv.data) == 0) {
	write("No structural variants detected.", file = 'sv_summary.tex', append = TRUE);
	} else {
	print(
		xtable(
			tool.summary,
			caption = 'Number of events (total and by SV type) discovered by each tool, as well as the overlap, and verified by MAVIS.'
			),
		file = 'sv_summary.tex',
		include.rownames = TRUE,
		append = TRUE
		);

	write("\\pagebreak\n\\begin{figure}[h!]", file = 'sv_summary.tex', append = TRUE);
	write("\\begin{center}", file = 'sv_summary.tex', append = TRUE);
	write(paste0(
		"\\includegraphics[width=0.9\\textwidth]{",
		getwd(), '/',
		generate.filename(arguments$project, 'SV_tool_overlap','png'), '}'
		), file = 'sv_summary.tex', append = TRUE);
	write("\\end{center}", file = 'sv_summary.tex', append = TRUE);
	write("\\caption{Number of structural variants detected by each tool or set of tools.}",
		file = 'sv_summary.tex', append = TRUE);
	write("\\end{figure}\n", file = 'sv_summary.tex', append = TRUE);

	if (length(normal.samples) > 1) {
		write("\\pagebreak\n\\begin{figure}[h!]", file = 'sv_summary.tex', append = TRUE);
		write("\\begin{center}", file = 'sv_summary.tex', append = TRUE);
		write(paste0(
			"\\includegraphics[width=0.9\\textwidth]{",
			getwd(), '/',
			generate.filename(arguments$project, 'germline_sv_summary','png'), '}'
			), file = 'sv_summary.tex', append = TRUE);
		write("\\end{center}", file = 'sv_summary.tex', append = TRUE);
		write("\\caption{Number of structural variants detected in germline samples, split by event type.}",
			file = 'sv_summary.tex', append = TRUE);
		write("\\end{figure}\n", file = 'sv_summary.tex', append = TRUE);
		}

	write("\\pagebreak\n\\begin{figure}[h!]", file = 'sv_summary.tex', append = TRUE);
	write("\\begin{center}", file = 'sv_summary.tex', append = TRUE);
	write(paste0(
		"\\includegraphics[width=0.9\\textwidth]{",
		getwd(), '/',
		generate.filename(arguments$project, 'somatic_sv_summary','png'), '}'
		), file = 'sv_summary.tex', append = TRUE);
	write("\\end{center}", file = 'sv_summary.tex', append = TRUE);
	write("\\caption{Number of structural variants detected in tumour samples, split by event type.}",
		file = 'sv_summary.tex', append = TRUE);
	write("\\end{figure}\n", file = 'sv_summary.tex', append = TRUE);

	if (nrow(event.counts) == 0) {
		write("No events were recurrently detected across the cohort.", file = 'sv_summary.tex', append = TRUE);
		} else {

		write("\\pagebreak\n\\subsection{Recurrent SVs}", file = 'sv_summary.tex', append = TRUE);

		for (type in names(recurrence.data)) {
			tmp <- xtable(
				recurrence.data[[type]],
				caption = paste0(
					'Recurrent ', type, ' events; all recurrent events (or top 10) are shown. Recurrent is defined as a given event type detected with identical breakpoints ($\\pm$1kbp) across multiple tumour samples.')
				);
			print(
				tmp,
				file = 'sv_summary.tex',
				include.rownames = FALSE,
				append = TRUE
				);
			}
		}
	}

### SAVE SESSION INFO ##############################################################################
save.session.profile(generate.filename('SV_Summary','SessionProfile','txt'));
