### plot_cn_mops_summary.R #########################################################################
# Plot CNA landscape
# INPUT:
#	- copy number calls (output by collect_panelCN_mops_output.R)

### FUNCTIONS ######################################################################################
# function to generate a standardized filename
generate.filename <- function(project.stem, file.core, extension, include.date = TRUE) {

	# build up the filename
	file.name <- paste(project.stem, file.core, sep = '_');
	file.name <- paste(file.name, extension, sep = '.');

	if (include.date) {
		file.name <- paste(Sys.Date(), file.name, sep = '_');
		}

	return(file.name);
	}

# function to write session profile to file
save.session.profile <- function(file.name) {

	# open the file
	sink(file = file.name, split = FALSE);

	# write memory usage to file
	cat('### MEMORY USAGE ###############################################################');
	print(proc.time());

	# write key variables to file
	cat("\n### VARIABLES #################################################################\n");
	cat(paste0('Input: ', arguments$cnas));
	cat(paste0('CNA type: ', arguments$scale));

	# write sessionInfo to file
	cat("\n### SESSION INFO ###############################################################");
	print(sessionInfo());

	# close the file
	sink();

	}

# function to trim sample IDs
simplify.ids <- function(x) {
	match <- TRUE;
	if (length(x) == 1) {
		match <- FALSE;
		new.ids <- x;
		}
	index <- 1;
	while (match) {
		if (length(unique(sapply(x,function(i) { unlist(strsplit(i,''))[index] } ))) == 1) {
			index <- index + 1;
			} else {
			new.ids <- sapply(x,function(i) { substring(i,index,nchar(i)) } );
			match <- FALSE;
			}
		}
	return(new.ids);
	}

### PREPARE SESSION ################################################################################
# import command line arguments
library(argparse);

parser <- ArgumentParser();

parser$add_argument('-p', '--project', type = 'character', help = 'PROJECT name');
parser$add_argument('-o', '--output', type = 'character', help = 'path to output directory');
parser$add_argument('-c', '--cnas', type = 'character',
	help = 'path to gene by sample matrix of cna calls (generated by collect_gatk_cnv_output.R)');
parser$add_argument('-s', '--scale', type = 'character',
	help = 'are we plotting CN or log2(ratio)? one of CN or ratio', default = 'CN');
parser$add_argument('-z', '--report', type = 'character', help = 'path to report directory',
	default = NULL);

arguments <- parser$parse_args();

# import libraries
library(BoutrosLab.plotting.general);
library(xtable);

### READ DATA ######################################################################################
# get data
if (is.null(arguments$cnas)) {
	stop('ERROR: No CNA calls provided, please provide path to CNA matrix from panelCN.mops.');
	} else {
	input.data <- read.delim(arguments$cnas, stringsAsFactors = FALSE);
	}

# create (if necessary) and move to output directory
if (!dir.exists(arguments$output)) {
	dir.create(arguments$output);
	}

setwd(arguments$output);

### FORMAT DATA ####################################################################################
# remove low quality calls
input.data <- input.data[which(input.data$lowQual != 'lowQual'),];

# format CN status
input.data$CN <- as.numeric(sub('CN','',as.character(input.data$CN))) - 2;

# get log2 ratio
input.data$LRR <- log2(input.data$RC.norm / input.data$medRC.norm);
input.data[which(input.data$CN == -2),]$LRR <- -2;

# sort data by pga data
input.data$Length <- input.data$End - input.data$Start;

pga <- merge(
	aggregate(Length ~ Sample, input.data, sum),
	aggregate(Length ~ Sample, input.data[which(input.data$CN != 0),], sum),
	by = 'Sample',
	suffixes = c('.total','.altered'),
	all = TRUE
	);

pga$PGA <- 100*(pga$Length.altered / pga$Length.total);
pga <- pga[order(pga$PGA, decreasing = TRUE),];

# get sample lists
all.samples <- unique(input.data$Sample);
sample.order <- rev(as.character(pga$Sample));

# makes sure input.data is sorted
input.data$Chromosome <- factor(input.data$Chr, levels = paste0('chr',c(1:22,'X','Y')));
input.data <- input.data[order(input.data$Chromosome, input.data$Start),];
input.data <- input.data[which(input.data$Chromosome != 'chrY'),];

# reshape for plotting
if (arguments$scale == 'CN') {
	plot.field <- 'CN';
	} else {
	plot.field <- 'LRR';
	}

cna.data <- reshape(
	input.data[,c('Chromosome','Start','End','Sample',plot.field)],
	direction = 'wide',
	timevar = 'Sample',
	idvar = c('Chromosome','Start','End')
	);

colnames(cna.data) <- gsub(paste0(plot.field,'.'),'',colnames(cna.data));
rownames(cna.data) <- paste0('Region',1:nrow(cna.data));

# trim out regions with mostly lowQual data (~5%)
region.counts <- apply(cna.data[,sample.order],1,function(i) { length(i[!is.na(i)])/length(i) } );
to.keep <- which(region.counts > 0.05);
cna.data <- cna.data[to.keep,];

# in case there is only 1 sample to plot
if (length(sample.order) == 1) {
	plot.data <- data.frame(cbind(cna.data, cna.data[,sample.order]));
	colnames(plot.data) <- rep(all.samples,2);
	} else {
	plot.data <- cna.data[,sample.order];
	}

### PLOT DATA ######################################################################################
# grab some parameters
axis.cex <- if (ncol(plot.data) <= 30) { 1
	} else if (ncol(plot.data) <= 50) { 0.75
	} else if (ncol(plot.data) <= 80) { 0.5
	} else if (ncol(plot.data) <= 100) { 0
	} else { 0 };

chromosomes <- gsub('chr','',tolower(cna.data$Chromosome));
chr.breaks <- get.line.breaks(chromosomes);
covariate.colours <- force.colour.scheme(chromosomes, scheme = 'chromosome');

chr.levels <- gsub('chr','',unique(tolower(cna.data$Chromosome)));

covariate.legends <- legend.grob(
	legends =  list(
		legend = list(
			colours = force.colour.scheme(chr.levels, scheme = 'chromosome'),
			labels = toupper(chr.levels)
			),
		legend = list(
			colours = 'grey50',
			labels = 'lowQual'
			)
		),
	size = 2
	);

# create heatmap
create.heatmap(
	plot.data,
	cluster.dimensions = 'none',
	covariates.top = list(
		rect = list(
			col = 'black',
			fill = covariate.colours,
			lwd = 0
			)
		),
	covariates.top.grid.border = list(col = 'black', lwd = 1),
#	covariates.top.grid.col = list(col = 'black', lwd = 1),
#	covariates.top.col.lines = chr.breaks,
	inside.legend = list(fun = covariate.legends, x = 1.02, y = 1),
	yaxis.lab = if (length(all.samples) == 1) { all.samples } else {
		gsub('\\.','-', simplify.ids(colnames(plot.data))) },
	yat = if (length(all.samples) == 1) { 1.5 } else { TRUE },
	xaxis.lab = rep('',nrow(plot.data)),
	yaxis.cex = axis.cex,
	xaxis.tck = 0,
	yaxis.tck = if (axis.cex == 0) { 0 } else { 0.2 },
	yaxis.fontface = 'plain',
	axes.lwd = 1,
#	col.colour = 'black',
#	grid.col = TRUE,
#	force.grid.col = TRUE,
#	col.lines = chr.breaks,
	print.colour.key = TRUE,
	fill.colour = 'grey50',
	at = if (arguments$scale == 'CN') { seq(-2.5,2.5,1) } else { seq(-2,2,0.1) },
	colour.scheme = c('blue','white','red'),
	colourkey.labels.at = if (arguments$scale == 'CN') { seq(-2,2,1) } else { seq(-2,2,1) },
	colourkey.labels = if (arguments$scale == 'CN') { seq(-2,2,1) } else { seq(-2,2,1) },
	colourkey.cex = 1,
	height = if (length(all.samples) > 12) { 8 } else { 5 },
	width = 12,
	resolution = 400,
	right.padding = 15,
	filename = generate.filename(arguments$project, 'panelCNMops_cna_landscape','png')
	);

save(
	cna.data,
	pga,
	file = generate.filename(arguments$project, 'panelCN_mops_data', 'RData')
	);

### LATEX ##########################################################################################
# if making output for a report
if (!is.null(arguments$report)) {

	tex.file <- paste0(
		arguments$report,
		'/cna_summary_mops.tex'
		);

	# create symlinks for plots
	unlink(paste0(arguments$report,'/panelCNMops_cna_landscape.png'));
	file.symlink(
		paste0(arguments$output,'/',
			generate.filename(arguments$project, 'panelCNMops_cna_landscape','png')),
		paste0(arguments$report,'/panelCNMops_cna_landscape.png')
		);

	# indicate caption to use with figure
	if (arguments$scale == 'CN') {
		cna.caption <- "Heatmap shows copy-number estimate; ordered by genomic position (chr1 [left] to chrX [right], indicated by top covariate) and percent target-bases copy-number altered (highest [top] to lowest [bottom]).";
		} else {
		cna.caption <- "Heatmap shows log$_2$(ratio); ordered by genomic position (chr1 [left] to chrX [right], indicated by top covariate) and percent target-bases copy-number altered (highest [top] to lowest [bottom]).";
		}

	# write latex for report
	write("\\section{SCNA Summary}", file = tex.file);
	write("Copy number variation was called using PanelCN.mops (v1.14.0) for all target regions:", file = tex.file, append = TRUE);

	write("\\begin{figure}[h!]", file = tex.file, append = TRUE);
	write("\\begin{center}", file = tex.file, append = TRUE);
	write(paste0(
		"\\includegraphics[width=0.9\\textwidth]{",
		paste0(arguments$report, '/', '/panelCNMops_cna_landscape.png'), '}'
		), file = tex.file, append = TRUE);
	write("\\end{center}", file = tex.file, append = TRUE);
	write(paste0(
		"\\caption{", cna.caption, "}"
		), file = tex.file, append = TRUE);
	write("\\end{figure}\n", file = tex.file, append = TRUE);
	}

### SAVE SESSION INFO ##############################################################################
save.session.profile(generate.filename('PanelCNmopsSummary','SessionProfile','txt'));
