#!/usr/bin/env perl
### bwa.pl #########################################################################################
use AutoLoader 'AUTOLOAD';
use strict;
use warnings;
use Carp;
use POSIX qw(strftime);
use Getopt::Std;
use Getopt::Long;
use File::Basename;
use File::Path qw(make_path);
use YAML qw(LoadFile);

my $cwd = dirname($0);
require "$cwd/shared/utilities.pl";

####################################################################################################
# version       author		comment
# 1.1           sprokopec	run BWA-MEM on raw FASTQ files (paired end only!)

#### USAGE ##########################################################################################
# bwa.pl -t tool_config.yaml -c data_config.yaml
#
# where:
# 	- tool_config.yaml contains tool versions and parameters, output directory,
# 	reference information, etc.
#	- data_config.yaml contains sample information (YAML file containing paths to FASTQ files,
#	generated by create_fastq_yaml.pl)

### SUBROUTINES ####################################################################################
# format command to generate readgroup
sub set_readgroup {
	my %args = (
		subject		=> undef,
		sample		=> undef,
		lane		=> undef,
		library		=> undef,
		platform	=> undef,
		@_
		);

	my $readgroup = '"@RG\tID:patient\tSM:smp\tPL:platform\tPU:unit\tLB:library"';
	$readgroup =~ s/patient/$args{subject}/;
	$readgroup =~ s/smp/$args{sample}/;
	$readgroup =~ s/unit/$args{lane}/;
	$readgroup =~ s/library/$args{library}/;
	$readgroup =~ s/platform/$args{platform}/;

	return($readgroup);
	}

# format command to run bwa then sort and index the output and remove tmp files
sub get_bwa_command {
	my %args = (
		r1		=> undef,
		r2		=> undef,
		stem		=> undef,
		readgroup	=> undef,
		reference	=> undef,
		@_
		);

	my $bwa_command = join(' ',
		'bwa mem -M -t4',  # 4 threads
		'-R', $args{readgroup},
		$args{reference},
		$args{r1}, $args{r2},
		'>', $args{stem} . '.sam'
		);

	return($bwa_command);
	}	

sub get_bwa_sort_command {
	my %args = (
		stem => undef,
		@_
		);

	my $sort_command = join(' ',
		'samtools sort -@4 -O bam',
		'-o', $args{stem} . '.bam',
		'-T', $args{stem}, $args{stem} . '.sam'
		);

	return($sort_command);
	}

sub get_index_bam_command {
	my %args = (
		output => undef,
		@_
		);

	my $index_command = join(' ',
		'samtools index',
		$args{output}
		);

	my $index_command2 = join(' ',
		'md5sum',
		$args{output},
		'>', $args{output} . '.md5'
		);
	
	my $command = join(";\n", $index_command, $index_command2);
	return($command);
	}

# format command to merge bams
sub get_merge_command {
	my %args = (
		input		=> undef,
		output		=> undef,
		tmp_dir		=> undef,
		java_mem	=> undef,
		@_
		);

	my $merge_command = join(' ',
		'java -Xmx' . $args{java_mem},
		'-jar $picard_dir/picard.jar MergeSamFiles',
		'INPUT=' . $args{input},
		'OUTPUT=' . $args{output},
		'ASSUME_SORTED=true USE_THREADING=true',
		'CREATE_INDEX=true CREATE_MD5_FILE=true',
		'TMP_DIR=' . $args{tmp_dir}
		);

	return($merge_command);
	}

sub get_merge_markdup_command {
	my %args = (
		input		=> undef,
		output		=> undef,
		tmp_dir		=> undef,
		java_mem	=> undef,
		@_
		);

	my $merge_command = join(' ',
		'java -Xmx' . $args{java_mem},
		'-jar $picard_dir/picard.jar MarkDuplicates',
		'INPUT=' . $args{input},
		'OUTPUT=' . $args{output},
		'METRICS_FILE=' . $args{output} . '.metrics',
		'TMP_DIR=' . $args{tmp_dir},
		'ASSUME_SORTED=true CREATE_INDEX=true CREATE_MD5_FILE=true',
		'MAX_RECORDS_IN_RAM=100000 VALIDATION_STRINGENCY=SILENT'
		);

	return($merge_command);
	}

sub main {
	my %args = (
		tool_config => undef,
		data_config => undef,
		@_
		);

	my $tool_config = $args{tool_config};
	my $data_config = $args{data_config};

	### PREAMBLE ######################################################################################

	# load tool config
	my $tool_data_orig = LoadFile($tool_config);
	my $tool_data = error_checking(tool_data => $tool_data_orig, pipeline => 'bwa');
	$tool_data->{date} = strftime "%F", localtime;

	# check for resume and confirm output directories
	my ($resume, $output_directory, $log_directory) = set_output_path(tool_data => $tool_data);

	# start logging
	print "---\n";
	print "Running BWA-MEM alignment pipeline for NGS data.\n";
	print "\n  Tool config used: $tool_config";
	print "\n    Reference: $tool_data->{reference}";
	print "\n    Output directory: $output_directory";
	print "\n  Sample config used: $data_config\n";
	print "\n---";

	# set tools and versions
	my $bwa_version = $tool_data->{tool} . '/' . $tool_data->{tool_version};
	my $samtools = 'samtools/' . $tool_data->{samtools_version};
	my $picard = 'picard/' . $tool_data->{picard_version};

	# create a file to hold job metrics
	my (@files, $run_count, $outfile, $touch_exit_status);
	if ('N' eq $tool_data->{dry_run}) {
		# initiate a file to hold job metrics (ensures that an existing file isn't overwritten by concurrent jobs)
		opendir(LOGFILES, $log_directory) or die "Cannot open $log_directory";
		@files = grep { /slurm_job_metrics/ } readdir(LOGFILES);
		$run_count = scalar(@files) + 1;
		closedir(LOGFILES);

		$outfile = $log_directory . '/slurm_job_metrics_' . $run_count . '.out';
		$touch_exit_status = system("touch $outfile");
		if (0 != $touch_exit_status) { Carp::croak("Cannot touch file $outfile"); }
		}

	### HANDLING FILES #################################################################################
	# get sample data
	my $smp_data = LoadFile($data_config);

	my ($run_script, $run_id, $run_id_extra, $raw_link, $final_link);
	my @all_jobs;

	# process each sample in $smp_data
	foreach my $patient (sort keys %{$smp_data}) {

		print "\nInitiating process for PATIENT: $patient\n";

		# make a sample-specific directory
		my $patient_directory = join('/', $output_directory, $patient);
		unless(-e $patient_directory) { make_path($patient_directory); }

		my @final_outputs;
		my $type;

		foreach my $sample (sort keys %{$smp_data->{$patient}}) {

			print "  SAMPLE: $sample\n";

			my $sample_directory = join('/', $patient_directory, $sample);
			unless(-e $sample_directory) { make_path($sample_directory); }

			my $tmp_directory = join('/', $sample_directory, 'TEMP');
			unless(-e $tmp_directory) { make_path($tmp_directory); }
			my $cleanup_cmd = "rm -rf $tmp_directory";

			my @libraries = keys %{$smp_data->{$patient}->{$sample}->{libraries}};
			my (@lane_intermediates, @lane_holds, @smp_jobs);

			# determine sample type
			$type = $smp_data->{$patient}->{$sample}->{type};

			foreach my $lib ( @libraries ) {

				print "    LIBRARY: $lib\n";

			#	my $lib_directory = join('/', $sample_directory, $lib);
			#	unless(-e $lib_directory) { make_path($lib_directory); }

				my @lanes = keys %{$smp_data->{$patient}->{$sample}->{libraries}->{$lib}->{runlanes}};

				foreach my $lane ( @lanes ) {

					print "      LANE: $lane\n";

				#	my $lane_directory = join('/', $lib_directory, $lane);
					my $lane_directory = join('/', $sample_directory, $lane);
					unless(-e $lane_directory) { make_path($lane_directory); }

					my $raw_directory = join('/', $lane_directory, 'fastq_links');
					unless(-e $raw_directory) { make_path($raw_directory); }

					# collect input files
					my $r1 = $smp_data->{$patient}->{$sample}->{libraries}->{$lib}->{runlanes}->{$lane}->{fastq}->{R1};
					my $r2 = $smp_data->{$patient}->{$sample}->{libraries}->{$lib}->{runlanes}->{$lane}->{fastq}->{R2};

					print "	R1: $r1\n";
					print "	R2: $r2\n\n";

					my @tmp = split /\//, $r1;
					$raw_link = join('/', $raw_directory, $tmp[-1]);
					symlink($r1, $raw_link);
					$raw_link =~ s/R1/R2/;
					symlink($r2, $raw_link);

					my $filestem = join('_', $sample, $lane);
					($run_id, $run_id_extra) = '';

					# run BWA-MEM on these fastqs
					my $readgroup = set_readgroup(
						subject		=> $patient,
						sample		=> $sample,
						lane		=> $lane,
						library 	=> $lib,
						platform 	=> $tool_data->{platform}
						);

					my $bwa = get_bwa_command(
						r1		=> $r1,
						r2		=> $r2,
						stem		=> $filestem,
						readgroup	=> $readgroup,
						reference	=> $tool_data->{reference}
						);

					$bwa = 'cd ' . $lane_directory . ";\n" . $bwa ;

					# check if this should be run
					if ( ('N' eq $resume) ||
						( 
							('Y' eq missing_file(join('/', $lane_directory, $filestem . '.sam'))) &&
							# if the index is missing, then this has been run before and the resulting sam deleted
							# (so don't do it again)
							('Y' eq missing_file(join('/', $lane_directory, $filestem . '.bam.bai')))
							)
						) {
						# record command (in log directory) and then run job
						print "Submitting job for bwa...\n";
						$run_script = write_script(
							log_dir => $log_directory,
							name	=> 'run_bwa_mem_' . $filestem,
							cmd	=> $bwa,
							modules => [$bwa_version, $samtools],
							max_time	=> $tool_data->{parameters}->{bwa}->{time}->{$type},
							mem		=> $tool_data->{parameters}->{bwa}->{mem}->{$type},
							cpus_per_task	=> 4,
							hpc_driver	=> $tool_data->{HPC_driver}
							);

						$run_id = submit_job(
							jobname		=> 'run_bwa_mem_' . $filestem,
							shell_command	=> $run_script,
							hpc_driver	=> $tool_data->{HPC_driver},
							dry_run		=> $tool_data->{dry_run}
							);

						push @smp_jobs, $run_id;
						push @all_jobs, $run_id;
						}
					else {
						print "Skipping alignment because this step was performed previously...\n";
						}

					# sort the resulting SAM and output BAM
					my $sort_cmd = get_bwa_sort_command(
						stem => $filestem
						);
					$sort_cmd = 'cd ' . $lane_directory . ";\n" . $sort_cmd;

					# check if this should be run
					if ( ('N' eq $resume) ||
						(
							('Y' eq missing_file(join('/', $lane_directory, $filestem . '.bam'))) &&
							# if the index is missing, then this has been run before and
							# the resulting sam deleted (so don't do it again)
							('Y' eq missing_file(join('/', $lane_directory, $filestem . '.bam.bai')))
							)
						) {
						# record command (in log directory) and then run job
						print "Submitting job to sort bam...\n"
;
						$run_script = write_script(
							log_dir	=> $log_directory,
							name	=> 'run_sort_sam_' . $filestem,
							cmd	=> $sort_cmd,
							modules	=> [$samtools],
							dependencies	=> $run_id,
							max_time	=> $tool_data->{parameters}->{sort}->{time}->{$type},
							mem		=> $tool_data->{parameters}->{sort}->{mem}->{$type},
							hpc_driver	=> $tool_data->{HPC_driver}
							);

						$run_id = submit_job(
							jobname		=> 'run_sort_sam_' . $filestem,
							shell_command	=> $run_script,
							hpc_driver	=> $tool_data->{HPC_driver},
							dry_run		=> $tool_data->{dry_run}
							);

						push @smp_jobs, $run_id;
						push @all_jobs, $run_id;
						}
					else {
						print "Skipping sort step because this was performed previously...\n";
						}

					# index the resulting BAM and remove intermediate SAM
					my $index_lane_cmd = get_index_bam_command(
						output => $filestem . '.bam'
						);
					$index_lane_cmd = 'cd ' . $lane_directory . ";\n" . $index_lane_cmd;

					$cleanup_cmd .= "\nrm $lane_directory/$filestem.sam";

					# check if this should be run
					if ( ('N' eq $resume) ||
						(
							('Y' eq missing_file(join('/', $lane_directory, $filestem . '.bam.bai'))) ||
							('Y' eq missing_file(join('/', $lane_directory, $filestem . '.bam.md5')))
							)
						) {
						# record command (in log directory) and then run job
						print "Submitting job to index bam...\n";

						$run_script = write_script(
							log_dir	=> $log_directory,
							name	=> 'run_bam_index_' . $filestem,
							cmd	=> $index_lane_cmd,
							modules	=> [$samtools],
							dependencies	=> $run_id,
							max_time	=> $tool_data->{parameters}->{index}->{time}->{$type},
							mem		=> $tool_data->{parameters}->{index}->{mem}->{$type},
							hpc_driver	=> $tool_data->{HPC_driver}
							);

						$run_id = submit_job(
							jobname		=> 'run_bam_index_' . $filestem,
							shell_command	=> $run_script,
							hpc_driver	=> $tool_data->{HPC_driver},
							dry_run		=> $tool_data->{dry_run}
							);

						push @smp_jobs, $run_id;
						push @all_jobs, $run_id;
						}
					else {
						print "Skipping index step because this was performed previously...\n";
						}

					push @lane_intermediates, $lane_directory . '/' . $filestem . '.bam';
					push @lane_holds, $run_id;
					push @smp_jobs, $run_id;
					push @all_jobs, $run_id;
					}
				}

			# if there are multiple libraries for a single sample, merge them
			my ($smp_output, $jobname, $merge_cmd);
			my $input_files = join(' INPUT=', @lane_intermediates);

			if ('N' eq $tool_data->{mark_dup}) {

				$smp_output = $patient_directory . '/' . $sample . '_bwamem_aligned_sorted_merged.bam';
				$jobname = 'run_merge_lanes_' . $sample;

				$merge_cmd = get_merge_command(
					input		=> $input_files,
					output		=> $smp_output,
					tmp_dir		=> $tmp_directory,
					java_mem	=> $tool_data->{parameters}->{merge}->{java_mem}->{$type}
					);
				}

			elsif ('Y' eq $tool_data->{mark_dup}) {

				$smp_output = $patient_directory . '/' . $sample . '_bwamem_aligned_sorted_merged_markdup.bam';
				$jobname = 'run_merge_markdup_' . $sample;

				$merge_cmd = get_merge_markdup_command(
					input		=> $input_files,
					output		=> $smp_output,
					tmp_dir		=> $tmp_directory,
					java_mem	=> $tool_data->{parameters}->{merge}->{java_mem}->{$type}
					);
				}

			$cleanup_cmd .= "\nrm " . join(";\nrm ", @lane_intermediates) . "\n";

			# check if this should be run
			if ( ('N' eq $resume) || ('Y' eq missing_file($smp_output)) ) {

				# IF THIS FINAL STEP IS SUCCESSFULLY RUN,
				# create a symlink for the final output in the TOP directory
				my @final = split /\//, $smp_output;
				$final_link = join('/', $tool_data->{output_dir}, $final[-1]);

				if (-l $final_link) {
					unlink $final_link or die "Failed to remove previous symlink: $final_link;\n";
					}

				my $link_cmd = "echo Creating new symlink...";
				$link_cmd .= "\nln -s $smp_output $final_link";

				$merge_cmd .= "\n" . $link_cmd;

				# record command (in log directory) and then run job
				print "Submitting job to merge lanes and mark dupilcates...\n";

				$run_script = write_script(
					log_dir	=> $log_directory,
					name	=> $jobname,
					cmd	=> $merge_cmd,
					modules	=> [$picard],
					dependencies	=> join(',', @lane_holds),
					mem 		=> $tool_data->{parameters}->{merge}->{mem}->{$type},
					max_time 	=> $tool_data->{parameters}->{merge}->{time}->{$type},
					hpc_driver	=> $tool_data->{HPC_driver}
					);

				$run_id = submit_job(
					jobname		=> $jobname,
					shell_command	=> $run_script,
					hpc_driver	=> $tool_data->{HPC_driver},
					dry_run		=> $tool_data->{dry_run}
					);

				push @smp_jobs, $run_id;
				push @all_jobs, $run_id;
				}
			else {
				print "Skipping mark dupicate step because this was performed previously...\n";	
				}

			push @final_outputs, $smp_output;

			# clean up/remove intermediate files (once per sample)
			if ('Y' eq $tool_data->{del_intermediate}) {

				print "Submitting job to clean up temporary/intermediate files...\n";

				# make sure final output exists before removing intermediate files!
				$cleanup_cmd = join("\n",
					"if [ -s $smp_output ]; then",
					$cleanup_cmd,
					"else",
					'echo "FINAL OUTPUT: $smp_output is missing; not removing intermediates"',
					"fi"
					);

				$run_script = write_script(
					log_dir	=> $log_directory,
					name	=> 'run_cleanup_' . $sample,
					cmd	=> $cleanup_cmd,
					dependencies	=> join(',', @smp_jobs),
					max_time	=> '00:05:00',
					mem		=> '256M',
					hpc_driver	=> $tool_data->{HPC_driver}
					);

				$run_id_extra = submit_job(
					jobname		=> 'run_cleanup_' . $sample,
					shell_command	=> $run_script,
					hpc_driver	=> $tool_data->{HPC_driver},
					dry_run		=> $tool_data->{dry_run}
					);
				}
			}

		print "FINAL OUTPUT:\n" . join("\n  ", @final_outputs) . "\n";
		print "---\n";
		}

	# if this is not a dry run, collect job metrics (exit status, mem, run time)
	if ('N' eq $tool_data->{dry_run}) {

		# collect job stats
		my $collect_metrics = collect_job_stats(
			job_ids	=> join(',', @all_jobs),
			outfile => $outfile
			);

		$run_script = write_script(
			log_dir	=> $log_directory,
			name	=> 'output_job_metrics_' . $run_count,
			cmd	=> $collect_metrics,
			dependencies	=> join(',', @all_jobs),
			max_time	=> '0:05:00',
			mem		=> '256M',
			hpc_driver	=> $tool_data->{HPC_driver}
			);

		$run_id = submit_job(
			jobname		=> 'output_job_metrics',
			shell_command	=> $run_script,
			hpc_driver	=> $tool_data->{HPC_driver},
			dry_run		=> $tool_data->{dry_run}
			);
		}

	# final job to output a BAM config for downstream stuff
	if ('Y' eq $tool_data->{create_output_yaml}) {

		print "Creating config yaml for output BAM files...\n";

		my $output_yaml_cmd = join(' ',
			"perl $cwd/shared/create_final_yaml.pl",
			'-d', $output_directory,
			'-o', $output_directory . '/bam_config.yaml',
			'-p', '.bam$' # *merged.bam|*markdup.bam
			);

		$run_script = write_script(
			log_dir	=> $log_directory,
			name	=> 'output_final_yaml',
			cmd	=> $output_yaml_cmd,
			modules	=> ['perl'],
			dependencies	=> join(',', @all_jobs),
			max_time	=> '0:10:00',
			mem		=> '1G',
			hpc_driver	=> $tool_data->{HPC_driver}
			);

		$run_id = submit_job(
			jobname		=> 'output_final_yaml',
			shell_command	=> $run_script,
			hpc_driver	=> $tool_data->{HPC_driver},
			dry_run		=> $tool_data->{dry_run}
			);

		} else {
			print "Not creating output config yaml as requested...\n";
		}

	# finish up
	print "\nProgramming terminated successfully.\n\n";

	}

### GETOPTS AND DEFAULT VALUES #####################################################################
# declare variables
my $tool_config;
my $data_config;

# read in command line arguments
GetOptions(
	't|tool=s'	=> \$tool_config,
	'c|config=s'    => \$data_config
	);

if (!defined($tool_config)) { die("No tool config file defined; please provide -t | --tool (ie, tool_config.yaml)"); }
if (!defined($data_config)) { die("No data config file defined; please provide -c | --config (ie, sample_config.yaml)"); }

main(tool_config => $tool_config, data_config => $data_config);
