#!/usr/bin/env perl
### pughlab_dnaseq_pipeline.pl #####################################################################
use AutoLoader 'AUTOLOAD';
use strict;
use warnings;
use Carp;
use POSIX qw(strftime);
use Getopt::Std;
use Getopt::Long;
use File::Basename;
use File::Path qw(make_path);
use YAML qw(LoadFile);
use List::Util 'any';

my $cwd = dirname($0);
require "$cwd/scripts/utilities.pl";

####################################################################################################
# version       author	  	comment
# 1.0		sprokopec       script to run PughLab DNASeq pipeline

### USAGE ##########################################################################################
# pughlab_dnaseq_pipeline.pl -c tool_config.yaml -d data.yaml
#
# where:
#	- tool_config.yaml contains tool versions and parameters, output directory, reference
#	information, etc.
#	- data_config.yaml contains sample information (YAML file containing paths to FASTQ files,
#	generated by create_fastq_yaml.pl)

### SUBROUTINES ####################################################################################

### MAIN ###########################################################################################
sub main {
	my %args = (
		tool_config	=> undef,
		data_config	=> undef,
		step1		=> undef,
		step2		=> undef,
		cleanup		=> undef,
		cluster		=> undef,
		dry_run		=> undef,
		@_
		);

	my $tool_config = $args{tool_config};
	my $data_config = $args{data_config};

	### PREAMBLE ######################################################################################

	# load tool config
	my $tool_data = LoadFile($tool_config);
	my $date = strftime "%F", localtime;
	my $timestamp = strftime "%F_%H-%M-%S", localtime;

	# check for and/or create output directories
	my $output_directory = $tool_data->{output_dir};
	$output_directory =~ s/\/$//;
	my $log_directory = join('/', $output_directory, 'logs', 'run_DNA_pipeline_' . $timestamp);

	unless(-e $output_directory) { make_path($output_directory); }
	unless(-e $log_directory) { make_path($log_directory); }

	# start logging
	my $log_file = join('/', $log_directory, 'run_DNASeq_pipeline.log');
	open (my $log, '>', $log_file) or die "Could not open $log_file for writing.";

	print $log "---\n";
	print $log "Running PughLab DNA-Seq pipeline.\n";
	print $log "\n  Tool config used: $tool_config";
	print $log "\n    Output directory: $output_directory";
	print $log "\n  Sample config used: $data_config";
	print $log "\n---\n\n";

	### MAIN ###########################################################################################

	my ($bwa_run_id, $gatk_run_id, $contest_run_id, $coverage_run_id, $hc_run_id) = '';
	my ($strelka_run_id, $mutect_run_id, $mutect2_run_id, $varscan_run_id, $delly_run_id) = '';
	my ($mavis_run_id) = '';

	# prepare directory structure
	my $bwa_directory = join('/', $output_directory, 'BWA');
	my $gatk_directory = join('/', $output_directory, 'GATK');
	my $contest_directory = join('/', $output_directory, 'BAMQC', 'ContEst');
	my $coverage_directory = join('/', $output_directory, 'BAMQC', 'Coverage');
	my $hc_directory = join('/', $output_directory, 'HaplotypeCaller');
	my $strelka_directory = join('/', $output_directory, 'Strelka');
	my $mutect_directory = join('/', $output_directory, 'MuTect');
	my $mutect2_directory = join('/', $output_directory, 'MuTect2');
	my $varscan_directory = join('/', $output_directory, 'VarScan');
	my $delly_directory = join('/', $output_directory, 'Delly');
	my $mavis_directory = join('/', $output_directory, 'Mavis');

	# indicate YAML files for processed BAMs
	my $bwa_output_yaml = join('/', $bwa_directory, 'bwa_bam_config_' . $timestamp . '.yaml');
	my $gatk_output_yaml = join('/', $gatk_directory, 'gatk_bam_config_' . $timestamp . '.yaml');

	if ( (!$args{step1}) && ($args{step2}) ) {
		$gatk_output_yaml = $data_config;
		$gatk_run_id = '000000';
		}

	# Should pre-processing (alignment + GATK indel realignment/recalibration + QC) be performed?
	if ($args{step1}) {

		## run BWA-alignment pipeline
		unless(-e $bwa_directory) { make_path($bwa_directory); }

		if (defined($tool_data->{bwa_config})) {

			my $bwa_command = join(' ',
				"perl $cwd/scripts/bwa.pl",
				"-o", $bwa_directory,
				"-t", $tool_data->{bwa_config},
				"-d", $data_config,
				"-b", $bwa_output_yaml,
				"-c", $args{cluster}
				);

			if ($args{cleanup}) {
				$bwa_command .= " --remove";
				}
			if ($args{dry_run}) {
				$bwa_command .= " --dry_run";
				}

			# record command (in log directory) and then run job
			print $log "Submitting job for bwa.pl\n";
			print $log "  COMMAND: $bwa_command\n\n";

			$bwa_run_id = `$bwa_command`;
			sleep(5);

			print $log ">>> Final BWA job id: $bwa_run_id\n\n";

			}

		## run GATK indel realignment/recalibration pipeline
		unless(-e $gatk_directory) { make_path($gatk_directory); }

		if (defined($tool_data->{gatk_config})) {

			my $gatk_command = join(' ',
				"perl $cwd/scripts/gatk.pl",
				"-o", $gatk_directory,
				"-t", $tool_data->{gatk_config},
				"-d", $bwa_output_yaml,
				"-c", $args{cluster},
				"-b", $gatk_output_yaml,
				"--depends", $bwa_run_id
				);

			if ($args{cleanup}) {
				$gatk_command .= " --remove";
				}
			if ($args{dry_run}) {
				$gatk_command .= " --dry_run";
				}

			# record command (in log directory) and then run job
			print $log "Submitting job for gatk.pl\n";
			print $log "  COMMAND: $gatk_command\n\n";

			$gatk_run_id = `$gatk_command`;
			sleep(5);

			print $log ">>> Final GATK job id: $gatk_run_id\n\n";

			}

		## run GATK's ContEst for contamination estimation (T/N only)
		## and GATK's DepthOfCoverage and find Callable Bases
		unless(-e $contest_directory) { make_path($contest_directory); }
		unless(-e $coverage_directory) { make_path($coverage_directory); }

		if (defined($tool_data->{bamqc_config})) {

			my $contest_command = join(' ',
				"perl $cwd/scripts/contest.pl",
				"-o", $contest_directory,
				"-t", $tool_data->{bamqc_config},
				"-d", $gatk_output_yaml,
				"-p", $tool_data->{project_name},
				"-c", $args{cluster},
				"--depends", $gatk_run_id
				);

			if ($args{cleanup}) {
				$contest_command .= " --remove";
				}
			if ($args{dry_run}) {
				$contest_command .= " --dry_run";
				}

			# record command (in log directory) and then run job
			print $log "Submitting job for contest.pl\n";
			print $log "  COMMAND: $contest_command\n\n";

			$contest_run_id = `$contest_command`;
			sleep(5);

			print $log ">>> Final ContEst job id: $contest_run_id\n\n";

			my $coverage_command = join(' ',
				"perl $cwd/scripts/get_coverage.pl",
				"-o", $coverage_directory,
				"-t", $tool_data->{bamqc_config},
				"-d", $gatk_output_yaml,
				"-p", $tool_data->{project_name},
				"-c", $args{cluster},
				"--depends", $gatk_run_id
				);

			if ($args{cleanup}) {
				$coverage_command .= " --remove";
				}
			if ($args{dry_run}) {
				$coverage_command .= " --dry_run";
				}

			# record command (in log directory) and then run job
			print $log "Submitting job for get_coverage.pl\n";
			print $log "  COMMAND: $coverage_command\n\n";

			$coverage_run_id = `$coverage_command`;
			sleep(5);

			print $log ">>> Final ContEst job id: $coverage_run_id\n\n";
			}
		}

	########################################################################################
	# From here on out, it makes more sense to run everything as a single batch.
	########################################################################################
	if ($args{step2}) {

		## run GATK's HaplotypeCaller pipeline
		if (defined($tool_data->{haplotype_caller_config})) {
	 
			my $hc_command = join(' ',
				"perl $cwd/scripts/haplotype_caller.pl",
				"-o", $hc_directory,
				"-t", $tool_data->{haplotype_caller_config},
				"-d", $gatk_output_yaml,
				"-c", $args{cluster},
				"--depends", $gatk_run_id
				);

			if ($args{cleanup}) {
				$hc_command .= " --remove";
				}
			if ($args{dry_run}) {
				$hc_command .= " --dry_run";
				}

			# record command (in log directory) and then run job
			print $log "Submitting job for haplotype_caller.pl\n";
			print $log "  COMMAND: $hc_command\n\n";

			$hc_run_id = `$hc_command`;
			sleep(5);

			print $log ">>> Final HaplotypeCaller job id: $hc_run_id\n\n";

			# next step will run Genotype GVCFs, and filter final output
			$hc_command = join(' ',
				"perl $cwd/scripts/genotype_gvcfs.pl",
				"-o", $hc_directory,
				"-t", $tool_data->{haplotype_caller_config},
				"-d", $gatk_output_yaml,
				"-p", $tool_data->{project_name},
				"-c", $args{cluster},
				"--depends", $hc_run_id
				);

			if ($args{cleanup}) {
				$hc_command .= " --remove";
				}
			if ($args{dry_run}) {
				$hc_command .= " --dry_run";
				}

			# record command (in log directory) and then run job
			print $log "Submitting job for genotype_gvcfs.pl\n";
			print $log "  COMMAND: $hc_command\n\n";
		
			$hc_run_id = `$hc_command`;
			sleep(5);
			}

		## run STRELKA/MANTA pipeline
		if (defined($tool_data->{strelka_config})) {

			unless(-e $strelka_directory) { make_path($strelka_directory); }

			# first create a panel of normals
			my $strelka_command = join(' ',
				"perl $cwd/scripts/strelka.pl",
				"-o", $strelka_directory,
				"-t", $tool_data->{strelka_config},
				"-d", $gatk_output_yaml,
				"-p", $tool_data->{project_name},
				"-c", $args{cluster},
				"--depends", $gatk_run_id
				);

			if ($args{cleanup}) {
				$strelka_command .= " --remove";
				}
			if ($args{dry_run}) {
				$strelka_command .= " --dry_run";
				}

			# record command (in log directory) and then run job
			print $log "Submitting job for strelka.pl\n";
			print $log "  COMMAND: $strelka_command\n\n";

			$strelka_run_id = `$strelka_command`;
			sleep(5);
			
			print $log ">>> Final Strelka job id: $strelka_run_id\n\n";
			}

		## run GATK's MuTect pipeline
		if (defined($tool_data->{mutect_config})) {

			unless(-e $mutect_directory) { make_path($mutect_directory); }

			# first create a panel of normals
			my $mutect_command = join(' ',
				"perl $cwd/scripts/mutect.pl",
				"-o", join('/', $mutect_directory, 'PanelOfNormals'),
				"-t", $tool_data->{mutect_config},
				"-d", $gatk_output_yaml,
				"-c", $args{cluster},
				"--depends", $gatk_run_id,
				"--create-panel-of-normals"
				);

			if ($args{cleanup}) {
				$mutect_command .= " --remove";
				}
			if ($args{dry_run}) {
				$mutect_command .= " --dry_run";
				}

			# record command (in log directory) and then run job
			print $log "Submitting job for mutect.pl --create-panel-of-normals\n";
			print $log "  COMMAND: $mutect_command\n\n";

			$mutect_run_id = `$mutect_command`;
			sleep(5);
			
			print $log ">>> Final MuTect PoN job id: $mutect_run_id\n\n";

			# next run variant calling
			$mutect_command = join(' ',
				"perl $cwd/scripts/mutect.pl",
				"-o", $mutect_directory,
				"-t", $tool_data->{mutect_config},
				"-d", $gatk_output_yaml,
				"-p", $tool_data->{project_name},
				"-c", $args{cluster},
				"--pon", join('/', $mutect_directory, 'panel_of_normals.vcf'),
				"--depends", $mutect_run_id
				);

			if ($args{cleanup}) {
				$mutect_command .= " --remove";
				}
			if ($args{dry_run}) {
				$mutect_command .= " --dry_run";
				}

			# record command (in log directory) and then run job
			print $log "Submitting job for mutect.pl\n";
			print $log "  COMMAND: $mutect_command\n\n";

			$mutect_run_id = `$mutect_command`;
			sleep(5);
			}

		## also run GATK's newer MuTect2 pipeline
		if (defined($tool_data->{mutect2_config})) {

			unless(-e $mutect2_directory) { make_path($mutect2_directory); }

			# first create a panel of normals
			my $mutect2_command = join(' ',
				"perl $cwd/scripts/mutect2.pl",
				"-o", join('/', $mutect2_directory, 'PanelOfNormals'),
				"-t", $tool_data->{mutect2_config},
				"-d", $gatk_output_yaml,
				"-c", $args{cluster},
				"--depends", $gatk_run_id,
				"--create-panel-of-normals"
				);

			if ($args{cleanup}) {
				$mutect2_command .= " --remove";
				}
			if ($args{dry_run}) {
				$mutect2_command .= " --dry_run";
				}

			# record command (in log directory) and then run job
			print $log "Submitting job for mutect2.pl --create-panel-of-normals\n";
			print $log "  COMMAND: $mutect2_command\n\n";

			$mutect2_run_id = `$mutect2_command`;
			sleep(5);
			
			print $log ">>> Final MuTect2 PoN job id: $mutect2_run_id\n\n";

			# now run the actual somatic variant calling
			$mutect2_command = join(' ',
				"perl $cwd/scripts/mutect2.pl",
				"-o", $mutect2_directory,
				"-t", $tool_data->{mutect2_config},
				"-d", $gatk_output_yaml,
				"-p", $tool_data->{project_name},
				"-c", $args{cluster},
				"--pon", join('/', $mutect2_directory, 'panel_of_normals.vcf'),
				"--depends", $mutect2_run_id
				);

			if ($args{cleanup}) {
				$mutect2_command .= " --remove";
				}
			if ($args{dry_run}) {
				$mutect2_command .= " --dry_run";
				}

			# record command (in log directory) and then run job
			print $log "Submitting job for mutect2.pl\n";
			print $log "  COMMAND: $mutect2_command\n\n";

			$mutect2_run_id = `$mutect2_command`;
			sleep(5);
			}

		## run VarScan SNV/CNV pipeline
		if (defined($tool_data->{varscan_config})) {

			unless(-e $varscan_directory) { make_path($varscan_directory); }

			# first run T/N pairs -- this will call germline variants for use in filtering T-only callsets
			my $varscan_command = join(' ',
				"perl $cwd/scripts/varscan.pl",
				"-o", $varscan_directory,
				"-t", $tool_data->{varscan_config},
				"-d", $gatk_output_yaml,
				"-p", $tool_data->{project_name},
				"-c", $args{cluster},
				"--mode paired",
				"--depends", $gatk_run_id
				);

			if ($args{cleanup}) {
				$varscan_command .= " --remove";
				}
			if ($args{dry_run}) {
				$varscan_command .= " --dry_run";
				}

			# record command (in log directory) and then run job
			print $log "Submitting job for varscan.pl (T/N mode)\n";
			print $log "  COMMAND: $varscan_command\n\n";

			$varscan_run_id = `$varscan_command`;
			sleep(5);
			
			print $log ">>> Final VarScan T/N job id: $varscan_run_id\n\n";

			# now run on T-only, using germline variants from T/N pairs to filter
			$varscan_command = join(' ',
				"perl $cwd/scripts/varscan.pl",
				"-o", $varscan_directory,
				"-t", $tool_data->{varscan_config},
				"-d", $gatk_output_yaml,
				"-p", $tool_data->{project_name},
				"-c", $args{cluster},
				"--mode unpaired",
				"--pon", join('/', $varscan_directory, 'panel_of_normals.vcf'),
				"--depends", $varscan_run_id
				);

			if ($args{cleanup}) {
				$varscan_command .= " --remove";
				}
			if ($args{dry_run}) {
				$varscan_command .= " --dry_run";
				}

			# record command (in log directory) and then run job
			print $log "Submitting job for varscan.pl\n";
			print $log "  COMMAND: $varscan_command\n\n";

			$varscan_run_id = `$varscan_command`;
			sleep(5);
			}

		## run Delly SV pipeline
		if (defined($tool_data->{delly_config})) {

			unless(-e $delly_directory) { make_path($delly_directory); }

			my $delly_command = join(' ',
				"perl $cwd/scripts/delly.pl",
				"-o", $delly_directory,
				"-t", $tool_data->{delly_config},
				"-d", $gatk_output_yaml,
				"-c", $args{cluster},
				"--depends", $gatk_run_id
				);

			if ($args{cleanup}) {
				$delly_command .= " --remove";
				}
			if ($args{dry_run}) {
				$delly_command .= " --dry_run";
				}

			# record command (in log directory) and then run job
			print $log "Submitting job for delly.pl\n";
			print $log "  COMMAND: $delly_command\n\n";

			$delly_run_id = `$delly_command`;
			sleep(5);

			print $log ">>> Final Delly job id: $delly_run_id\n\n";
			}

		## run Mavis SV annotation pipeline
		if (defined($tool_data->{mavis_config})) {

			unless(-e $mavis_directory) { make_path($mavis_directory); }

			my $mavis_command = join(' ',
				"perl $cwd/scripts/mavis.pl",
				"-o", $mavis_directory,
				"-t", $tool_data->{mavis_config},
				"-d", $gatk_output_yaml,
				"--manta", $strelka_directory,
				"--delly", $delly_directory,
				"-c", $args{cluster}
				);

			if ($args{cleanup}) {
				$mavis_command .= " --remove";
				}
			if ($args{dry_run}) {
				$mavis_command .= " --dry_run";
				}

			# record command (in log directory) and then run job
			print $log "Submitting job for mavis.pl\n";
			print $log "  COMMAND: $mavis_command\n\n";

			# because mavis.pl will search provided directories and run based on what it finds
			# (Manta/Delly resuts), this can only be run AFTER these respective jobs finish
			# so, we will submit this with dependencies
			my $run_script = write_script(
				log_dir	=> $log_directory,
				name	=> 'pughlab_dna_pipeline__run_mavis',
				cmd	=> $mavis_command,
				modules	=> ['perl'],
				dependencies	=> join(',', $delly_run_id, $strelka_run_id),
				mem		=> '256M',
				hpc_driver	=> $args{cluster}
				);

			$mavis_run_id = submit_job(
				jobname		=> $log_directory,
				shell_command	=> $run_script,
				hpc_driver	=> $args{cluster},
				dry_run		=> $args{dry_run},
				log_file	=> $log
				);

			print $log ">>> MAVIS job id: $mavis_run_id\n\n";
			}
		}

	# finish up
	print $log "\nProgramming terminated successfully.\n\n";
	close $log;

	}

### GETOPTS AND DEFAULT VALUES #####################################################################
# declare variables
my ($tool_config, $data_config);
my ($preprocessing, $variant_calling);
my $hpc_driver = 'slurm';
my ($remove_junk, $dry_run);
my $help;

# read in command line arguments
GetOptions(
	'h|help'		=> \$help,
	't|tool=s'		=> \$tool_config,
	'd|data=s'		=> \$data_config,
	'preprocessing'		=> \$preprocessing,
	'variant_calling'	=> \$variant_calling,
	'c|cluster=s'		=> \$hpc_driver,
	'remove'		=> \$remove_junk,
	'dry_run'		=> \$dry_run,
	 );

if ($help) {
	my $help_msg = join("\n",
		"Options:",
		"\t--help|-h\tPrint this help message",
		"\t--data|-d\t<string> data config (yaml format)",
		"\t--tool|-t\t<string> tool config (yaml format)",
		"\t--preprocessing\t<boolean> should data pre-processing be performed? (default: false)",
		"\t--variant_calling\t<boolean> should variant calling be performed? (default: false)",
		"\t--cluster|-c\t<string> cluster scheduler (default: slurm)",
		"\t--remove\t<boolean> should intermediates be removed? (default: false)",
		"\t--dry_run\t<boolean> should jobs be submitted? (default: false)"
		);

	print $help_msg;
	exit;
	}

if ( (!$preprocessing) && (!$variant_calling) ) {
	die("Please choose a step to run (either --preprocessing and/or --variant_caling)");
	}
if (!defined($tool_config)) { die("No tool config file defined; please provide -t | --tool (ie, tool_config.yaml)"); }
if (!defined($data_config)) { die("No data config file defined; please provide -d | --data (ie, sample_config.yaml)"); }

# check for compatible HPC driver; if not found, change dry_run to Y
my @compatible_drivers = qw(slurm);
if ( (!any { /$hpc_driver/ } @compatible_drivers ) && (!$dry_run) ) {
	print "Unrecognized HPC driver requested: setting dry_run to true -- jobs will not be submitted but commands will be written to file.\n";
	$dry_run = 1;
	}

main(
	tool_config	=> $tool_config,
	data_config	=> $data_config,
	step1		=> $preprocessing,
	step2		=> $variant_calling,
	cluster		=> $hpc_driver,
	cleanup		=> $remove_junk,
	dry_run		=> $dry_run
	);
