### collect_panelCN_mops_output.R ##################################################################
# Finds and combines output generated by PanelCN.mops

### FUNCTIONS ######################################################################################
# function to generate a standardized filename
generate.filename <- function(project.stem, file.core, extension, include.date = TRUE) {

	# build up the filename
	file.name <- paste(project.stem, file.core, sep = '_');
	file.name <- paste(file.name, extension, sep = '.');

	if (include.date) {
		file.name <- paste(Sys.Date(), file.name, sep = '_');
		}

	return(file.name);
	}

# function to write session profile to file
save.session.profile <- function(file.name) {

	# open the file
	sink(file = file.name, split = FALSE);

	# write memory usage to file
	cat('### MEMORY USAGE ###############################################################');
	print(proc.time());

	# write sessionInfo to file
	cat("\n### SESSION INFO ###############################################################");
	print(sessionInfo());

	# close the file
	sink();

	}

### PREPARE SESSION ################################################################################
# import libraries
library(argparse);

# import command line arguments
parser <- ArgumentParser();

parser$add_argument('-d', '--directory', type = 'character', help = 'path to data directory');
parser$add_argument('-p', '--project', type = 'character', help = 'project name');

arguments <- parser$parse_args();

# what's the date?
date <- Sys.Date();

setwd(arguments$directory);

### MAIN ###########################################################################################
# find results files
cn.files <- list.files(pattern = 'panelcn.mops_results.tsv$', recursive = TRUE);

# read them in
if (any(grepl('cohort',cn.files))) {
	cohort.data <- read.delim(cn.files[grepl('cohort',cn.files)]);
	cn.files <- cn.files[!grepl('cohort',cn.files)];
	} else {
	cohort.data <- NULL;
	}

cn.data <- do.call(rbind,
	lapply(
		cn.files,
		read.delim
		)
	);

# save raw data to file
write.table(
	cn.data,
	file = generate.filename(arguments$project, 'panelCN.mops_output','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

# compare individual runs with cohort run if present
if (!is.null(cohort.data)) {

	key.fields <- c('Sample','Chr','Start','End','Gene','CN');
	tmp <- merge(
		cn.data[,key.fields],
		cohort.data[,key.fields],
		by = c('Sample','Chr','Start','End','Gene'),
		suffixes = c('.split','.cohort')
		);
	tmp$CN.split <- as.numeric(gsub('CN','',as.character(tmp$CN.split)));
	tmp$CN.cohort <- as.numeric(gsub('CN','',as.character(tmp$CN.cohort)));

	print(paste0(
		"Spearman's correlation between individual runs and cohort run (CN values) is: ",
		cor(tmp$CN.split,tmp$CN.cohort, use = 'pairwise', method = 'spearman')
		));

	cor.results <- data.frame(Sample = unique(tmp$Sample), Correlation = NA);
	for (i in 1:nrow(cor.results)) {
		smp <- cor.results[i,]$Sample;
		cor.results[i,]$Correlation <- cor(
			tmp[which(tmp$Sample == smp),]$CN.split,
			tmp[which(tmp$Sample == smp),]$CN.cohort,
			use = 'pairwise',
			method = 'spearman'
			);
		}

	print(paste0(
		"Per-sample Spearman's correlations between CN values range from ",
		min(cor.results$Correlation), " to ", max(cor.results$Correlation)
		));

	# save raw data to file
	write.table(
		cor.results
		file = generate.filename(arguments$project, 
			'split_vs_combined_per-sample_correlations','tsv'),
		row.names = FALSE,
		col.names = TRUE,
		sep = '\t'
		);
	}

# format data
cn.data$CN <- as.numeric(gsub('CN','',as.character(cn.data$CN)));
cn.data[which(cn.data$lowQual == 'lowQual'),]$CN <- NA;
colnames(cn.data)[2] <- 'Chromosome';

# reshape discrete CN calls
cn.calls <- reshape(
	cn.data[,c('Chromosome','Gene','Exon','Start','End','Sample','CN')],
	direction = 'wide',
	idvar = c('Chromosome','Gene','Exon','Start','End'),
	timevar = 'Sample'
	);
colnames(cn.calls) <- gsub('CN.','',colnames(cn.calls));

# calculate log2ratios
cn.data$Ratio <- log2(cn.data$RC.norm / cn.data$medRC.norm);

# set minimum ratio to -2 (to deal with -Inf)
cn.data[which(cn.data$CN == 0),]$Ratio <- -2;

ratio.data <- reshape(
	cn.data[,c('Chromosome','Gene','Exon','Start','End','Sample','Ratio')],
	direction = 'wide',
	idvar = c('Chromosome','Gene','Exon','Start','End'),
	timevar = 'Sample'
	);
colnames(ratio.data) <- gsub('Ratio.','',colnames(ratio.data));

# save combined/formatted data to file
write.table(
	cn.calls,
	file = generate.filename(arguments$project, 'panelCN.mops_cna_matrix','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

write.table(
	ratio.data,
	file = generate.filename(arguments$project, 'panelCN.mops_ratio_matrix','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

### SAVE SESSION INFO ##############################################################################
save.session.profile(generate.filename('CollectPanelCNmops','SessionProfile','txt'));
