### collect_sequenza_output.R ######################################################################
# Finds and combines output generated by Sequenza (CN segments, cellularity, ploidy)

### FUNCTIONS ######################################################################################
# function to generate a standardized filename
generate.filename <- function(project.stem, file.core, extension, include.date = TRUE) {

	# build up the filename
	file.name <- paste(project.stem, file.core, sep = '_');
	file.name <- paste(file.name, extension, sep = '.');

	if (include.date) {
		file.name <- paste(Sys.Date(), file.name, sep = '_');
		}

	return(file.name);
	}

# function to write session profile to file
save.session.profile <- function(file.name) {

	# open the file
	sink(file = file.name, split = FALSE);

	# write memory usage to file
	cat('### MEMORY USAGE ###############################################################');
	print(proc.time());

	# write sessionInfo to file
	cat("\n### SESSION INFO ###############################################################");
	print(sessionInfo());

	# close the file
	sink();

	}

### PREPARE SESSION ################################################################################
# import libraries
library(GenomicRanges);
library(argparse);

# import command line arguments
parser <- ArgumentParser();

parser$add_argument('-d', '--directory', type = 'character', help = 'path to data directory');
parser$add_argument('-p', '--project', type = 'character', help = 'project name');
parser$add_argument('-g', '--gtf', type = 'character', help = 'annotation gtf (refGene)',
	default = '/cluster/projects/pughlab/references/gencode/GRCh38/gencode.v31.annotation.gtf');
parser$add_argument('-t', '--targets', type = 'character', help = 'target intervals');

arguments <- parser$parse_args();

# what's the date?
date <- Sys.Date();

setwd(arguments$directory);

### FORMAT ANNOTATION
# using refGene, indicate genes to keep (coding genes) for gene x patient matrix
refGene <- read.delim(arguments$gtf, header = F, comment.char = '#');

refGene <- droplevels(refGene[which(refGene$V3 == 'gene'),c(1,4,5,9)]);
refGene <- droplevels(refGene[which(refGene$V1 != 'chrM'),]);
colnames(refGene) <- c('Chromosome','Start','End','INFO');

refGene <- refGene[grepl('gene_type protein_coding', refGene$INFO),];
refGene <- refGene[!grepl('tag PAR', refGene$INFO),];

gene.widths <- refGene$End - refGene$Start;
refGene <- droplevels(refGene[which(gene.widths > 20),]);

refGene.gr <- GRanges(refGene);

refGene$GeneID <- sapply(
        refGene$INFO,
        function(i) {
                parts <- unlist(strsplit(as.character(i), ';'));
                gene_name <- unlist(strsplit(parts[grepl('gene_id', parts)], ' '));
                gene_id  <- gene_name[length(gene_name)];
                return(gene_id);
                }
        );

refGene$Symbol <- sapply(
        refGene$INFO,
        function(i) {
                parts <- unlist(strsplit(as.character(i), ';'));
                gene_name <- unlist(strsplit(parts[grepl('gene_name', parts)], ' '));
                gene_symbol <- gene_name[length(gene_name)];
                return(gene_symbol);
                }
        );

refGene$Chromosome <- factor(refGene$Chromosome, levels = paste0('chr',c(1:22,'X','Y')));
refGene <- refGene[,-4];

gene.gr <- GRanges(refGene);

# if provided, filter to target intervals
refGene$Target <- NA;
if (!is.null(arguments$targets)) {

	target.intervals <- read.delim(arguments$targets, header = F, comment.char = '@');
	colnames(target.intervals)[1:4] <- c('Chromosome','Start','End','Strand');

	target.gr <- GRanges(target.intervals);

	overlaps <- as.data.frame(findOverlaps(refGene.gr, target.gr));

	refGene$Target <- 0;
	refGene[unique(overlaps$queryHits),]$Target <- 1;
	}

### MAIN ###########################################################################################
# find results files
cr.files <- list.files(pattern = 'called.seg', recursive = TRUE);
#cr.files <- cr.files[!grepl('^2021', cr.files)];

# read them in
cn.list <- list();

for (file in cr.files) {
	# extract sample ID
	smp <- unlist(strsplit(basename(file), '\\.'))[1];
	# store data in list
	tmp <- read.delim(file, as.is = TRUE, comment.char = '@');
	tmp$ID <- smp;
	cn.list[[smp]] <- tmp;
	rm(tmp);
	}

cn.data <- do.call(rbind, cn.list);

# adjust slightly for compatibility with cbioportal
cn.data <- cn.data[,c('ID','CONTIG','START','END','NUM_POINTS_COPY_RATIO','MEAN_LOG2_COPY_RATIO','CALL')];
colnames(cn.data) <- c('ID','chrom','loc.start','loc.end','num.mark','seg.mean','call');

# save combined/formatted data to file
write.table(
	cn.data[,1:6],
	file = generate.filename(arguments$project, 'gatk_cnv_for_gistic', 'tsv'),
	row.names = FALSE,
	col.names = TRUE,
	quote = FALSE,
	sep = '\t'
	);

cn.data$chrom <- gsub('chr','',cn.data$chrom);
write.table(
	cn.data,
	file = generate.filename(arguments$project, 'gatk_cnv_for_cbioportal', 'seg'),
	row.names = FALSE,
	col.names = TRUE,
	quote = FALSE,
	sep = '\t'
	);

# define a results table
gene.data.cn <- refGene;
for (smp in names(cn.list)) {
	gene.data.cn[,smp] <- 0;
	}
gene.data.ratio <- gene.data.cn;

pga.values <- list();

for (smp in names(cn.list)) {

	# convert to GRanges
	tmp <- cn.list[[smp]][,c('CONTIG','START','END','MEAN_LOG2_COPY_RATIO','CALL')];
	colnames(tmp) <- c('chrom','start','end','RATIO','CN');
	tmp$CN <- factor(tmp$CN, levels = c('-','0','+'), labels = c(-1,0,1));
	tmp$CN <- as.numeric(as.character(tmp$CN));

	if (nrow(tmp) == 0) { next; }
	rownames(tmp) <- paste0('seg', 1:nrow(tmp));

	# calculate PGA
	for.pga <- tmp[which(tmp$CN != 0),];
	pga <- sum(for.pga$end - for.pga$start)/(3*10**9)*100;
	pga.values[[smp]] <- pga;

	# extract gene data
	gr <- GRanges(tmp);

	# find overlap with genes (any amount of overlap)
	overlapGenes <- as.data.frame(findOverlaps(gr, gene.gr, minoverlap = 20));
	regionsWithHits <- as.data.frame(gr[unique(overlapGenes$queryHits)]);

	for (i in rownames(regionsWithHits)) {
		genes <- gene.gr[overlapGenes[which(overlapGenes$queryHits == sub('seg','',i)),]$subjectHits,]$GeneID;
		gene.data.cn[which(gene.data.cn$GeneID %in% genes),smp] <- tmp[i,]$CN;
		gene.data.ratio[which(gene.data.ratio$GeneID %in% genes),smp] <- tmp[i,]$RATIO;
		}
	}

pga.data <- do.call(rbind, pga.values);
colnames(pga.data)[1] <- 'PGA';

# save data to file
write.table(
	pga.data,
	file = generate.filename(arguments$project, 'gatk_pga_estimates','tsv'),
	row.names = TRUE,
	col.names = TRUE,
	sep = '\t'
	);

colnames(gene.data.cn)[which(colnames(gene.data.cn) == 'Symbol')] <- 'Hugo_Symbol';
write.table(
	gene.data.cn[,c('Hugo_Symbol',names(cn.list))],
	file = generate.filename(arguments$project, 'gatk_cna_matrix_for_cbioportal', 'tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

write.table(
	gene.data.ratio,
	file = generate.filename(arguments$project, 'gatk_ratio_gene_matrix', 'tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

### SAVE SESSION INFO ##############################################################################
save.session.profile(generate.filename('CollectCNAs','SessionProfile','txt'));
