### collect_mavis_output.R #########################################################################
# Finds and combines output generated by MAVIS.

### FUNCTIONS ######################################################################################
# function to generate a standardized filename
generate.filename <- function(project.stem, file.core, extension, include.date = TRUE) {

	# build up the filename
	file.name <- paste(project.stem, file.core, sep = '_');
	file.name <- paste(file.name, extension, sep = '.');

	if (include.date) {
		file.name <- paste(Sys.Date(), file.name, sep = '_');
		}

	return(file.name);
	}

# function to write session profile to file
save.session.profile <- function(file.name) {

	# open the file
	sink(file = file.name, split = FALSE);

	# write memory usage to file
	cat('### MEMORY USAGE ###############################################################');
	print(proc.time());

	# write sessionInfo to file
	cat("\n### SESSION INFO ###############################################################");
	print(sessionInfo());

	# close the file
	sink();

	}

### PREPARE SESSION ################################################################################
# import libraries
library(argparse);
library(plyr);

# import command line arguments
parser <- ArgumentParser();

parser$add_argument('-d', '--directory', type = 'character', help = 'path to data directory');
parser$add_argument('-p', '--project', type = 'character', help = 'project name');

arguments <- parser$parse_args();

# what's the date?
date <- Sys.Date();

setwd(arguments$directory);

### MAIN ###########################################################################################
# find results files
mavis.files <- list.files(pattern = 'mavis_summary_all', recursive = TRUE);

# read them in
sv.list <- list();

for (file in mavis.files) {
	# extract sample ID
	smp <- unlist(strsplit(file, '\\/'))[1];
	# read in and store data in list
	sv.list[[smp]] <- read.delim(file, as.is = TRUE);
	}

# reshape/format data
sv.data.full <- join_all(
	sv.list,
	type = 'full',
	match = 'first'
	);
colnames(sv.data.full)[which(colnames(sv.data.full) == 'X.tracking_id')] <- 'tracking_id';

key.fields <- c('event_type','gene1_aliases','gene2_aliases','fusion_splicing_pattern','break1_chromosome','break1_position_start','break1_position_end','break2_chromosome','break2_position_start','break2_position_end','break1_split_reads','break2_split_reads','spanning_reads','flanking_pairs','linking_split_reads','library','tracking_id','tools');

smp.fields <- sort(colnames(sv.data.full)[grep('_genome',colnames(sv.data.full))]);

sv.data <- sv.data.full[,c(key.fields, smp.fields)];

# save combined/formatted data to file
write.table(
	sv.data,
	file = generate.filename(arguments$project, 'mavis_output','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

# identify recurrent* hits (gene-wise events)
genes.tmp <- unique(sv.data[,c('gene1_aliases','gene2_aliases','event_type','library')]);
genes.tmp <- genes.tmp[-which(genes.tmp$gene1_aliases == 'None' & genes.tmp$gene2_aliases == 'None'),];

gene.data <- reshape(
	genes.tmp,
	direction = 'wide',
	timevar = 'library',
	idvar = c('gene1_aliases','gene2_aliases')
	);
colnames(gene.data) <- gsub('event_type.','',colnames(gene.data));
colnames(gene.data)[1:2] <- c('gene1','gene2');

# because some samples may have multiple events within the same gene-gene pair
smp.counts <- aggregate(event_type ~ gene1_aliases + gene2_aliases + library, genes.tmp, length);
smp.counts <- smp.counts[which(smp.counts$event_type == 2),];

for (i in 1:nrow(smp.counts)) {

	gene1 <- smp.counts[i,1];
	gene2 <- smp.counts[i,2];
	smp <- smp.counts[i,3];

	idx <- which(genes.tmp$gene1_aliases == gene1 & genes.tmp$gene2_aliases == gene2 & genes.tmp$library == smp);
	types <- genes.tmp[idx,]$event_type;

	gene.data[which(gene.data$gene1 == gene1 & gene.data$gene2 == gene2),smp] <- paste(types, collapse = ',');

	}

# save combined/formatted data to file
write.table(
	gene.data,
	file = generate.filename(arguments$project, 'gene_by_sample','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

### SAVE SESSION INFO ##############################################################################
save.session.profile(generate.filename('CollectMAVIS','SessionProfile','txt'));
