### collect_mavis_output.R #########################################################################
# Finds and combines output generated by MAVIS.

### FUNCTIONS ######################################################################################
# function to generate a standardized filename
generate.filename <- function(project.stem, file.core, extension, include.date = TRUE) {

	# build up the filename
	file.name <- paste(project.stem, file.core, sep = '_');
	file.name <- paste(file.name, extension, sep = '.');

	if (include.date) {
		file.name <- paste(Sys.Date(), file.name, sep = '_');
		}

	return(file.name);
	}

# function to write session profile to file
save.session.profile <- function(file.name) {

	# open the file
	sink(file = file.name, split = FALSE);

	# write memory usage to file
	cat('### MEMORY USAGE ###############################################################');
	print(proc.time());

	# write sessionInfo to file
	cat("\n### SESSION INFO ###############################################################");
	print(sessionInfo());

	# close the file
	sink();

	}

### PREPARE SESSION ################################################################################
# import libraries
library(argparse);
library(plyr);

# import command line arguments
parser <- ArgumentParser();

parser$add_argument('-d', '--directory', type = 'character', help = 'path to data directory');
parser$add_argument('-p', '--project', type = 'character', help = 'project name');

arguments <- parser$parse_args();

# what's the date?
date <- Sys.Date();

setwd(arguments$directory);

### MAIN ###########################################################################################
# find results files
mavis.files <- list.files(pattern = 'mavis_summary_all', recursive = TRUE);

# read them in
sv.list <- list();

for (file in mavis.files) {
	# extract sample ID
	smp <- unlist(strsplit(file, '\\/'))[1];
	# read in and store data in list
	sv.list[[smp]] <- read.delim(file, as.is = TRUE);
	}

# reshape/format data
sv.data.full <- join_all(
	sv.list,
	type = 'full',
	match = 'first'
	);
colnames(sv.data.full)[which(colnames(sv.data.full) == 'X.tracking_id')] <- 'tracking_id';

key.fields <- c('event_type','gene1_aliases','gene2_aliases','fusion_splicing_pattern','break1_chromosome','break1_position_start','break1_position_end','break2_chromosome','break2_position_start','break2_position_end','break1_split_reads','break2_split_reads','spanning_reads','flanking_pairs','linking_split_reads','library','tracking_id','tools');

smp.fields <- sort(colnames(sv.data.full)[grep('_genome|_transcriptome',colnames(sv.data.full))]);
smp.fields.names <- sapply(smp.fields, function(i) { unlist(strsplit(i,'_'))[1] } );

sv.data <- sv.data.full[,c(key.fields, smp.fields)];

# save combined/formatted data to file
write.table(
	sv.data,
	file = generate.filename(arguments$project, 'mavis_output','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

# format for cbioportal (not complete!)
tmp <- sv.data[,key.fields];
tmp$Fusion <- paste0(tmp$gene1_aliases, '--', tmp$gene2_aliases);
tmp <- tmp[!grepl('None',tmp$Fusion),];

tmp$Status <- 'unknown';
for (i in 1:nrow(tmp)) {
	smp <- gsub('-','\\.',tmp[i,]$library);
	title <- smp.fields[which(smp.fields.names == smp)];
	if (grepl('diseased', title)) { tmp[i,]$Status <- 'somatic'; }
	else if (grepl('normal', title)) { tmp[i,]$Status <- 'germline'; }
	}

tmp$Frame <- 'unknown';
tmp[which(tmp$fusion_splicing_pattern == 'normal'),]$Frame <- 'inframe';
tmp[which(!tmp$fusion_splicing_pattern %in% c('normal','None')),]$Frame <- 'frameshift';

tmp$Support <- NA;
for (i in 1:nrow(tmp)) {
	type <- if (grepl('rna', tmp[i,]$library)) { 'rna' } else { 'dna' }
	tmp[i,]$Support <- type;
	}

# aggregate to get unique entries
unique.calls <- aggregate(
	tmp$tools,
	by = list(
		gene1_aliases = tmp$gene1_aliases,
		gene2_aliases = tmp$gene2_aliases,
		library = tmp$library,
		event_type = tmp$event_type,
		Status = tmp$Status,
		Fusion = tmp$Fusion,
		Frame = tmp$Frame,
		Support = tmp$Support
		),
	FUN = function(i) {
		paste(c(unique(unlist(strsplit(as.character(i),';'))),'mavis'), collapse = ',')
		}
	);

colnames(unique.calls)[which(colnames(unique.calls) == 'x')] <- 'Tool';

unique.calls$Sample <- gsub('-wgs|-wxs|-rna','',unique.calls$library);

# format for cbioportal
cbio.data <- NULL;
dna.fusions <- NULL;
if (nrow(unique.calls[which(unique.calls$Support == 'dna'),]) > 0) {
	dna.fusions <- unique.calls[which(unique.calls$Support == 'dna'),];

	cbio.data <- data.frame(
		Hugo_Symbol = c(
			dna.fusions$gene1_aliases,
			dna.fusions$gene2_aliases
			),
		Entrez_Gene_Id = NA,
		Center = NA,
		Tumor_Sample_Barcode = rep(dna.fusions$Sample, times = 2),
		Fusion = rep(dna.fusions$Fusion, times = 2),
		DNA_support = 'yes',
		RNA_support = 'no',
		Method = rep(dna.fusions$Tool, times = 2),
		Frame = rep(dna.fusions$Frame, times = 2),
		Fusion_Status = rep(dna.fusions$Status, times = 2),
#		SV_Type = rep(dna.fusions$event_type, times = 2),
#		Seq_Type = rep(dna.fusions$Seq_Type, times = 2),
		stringsAsFactors = FALSE
		);
	}

rna.fusions <- NULL;
if (nrow(unique.calls[which(unique.calls$Support == 'rna'),]) > 0) {
	rna.fusions <- unique.calls[which(unique.calls$Support == 'rna'),];

	rna.cbio <- data.frame(
		Hugo_Symbol = c(
			rna.fusions$gene1_aliases,
			rna.fusions$gene2_aliases
			),
		Entrez_Gene_Id = NA,
		Center = NA,
		Tumor_Sample_Barcode = gsub('.rna','',rep(rna.fusions$Sample, times = 2)),
		Fusion = rep(rna.fusions$Fusion, times = 2),
		DNA_support = 'no',
		RNA_support = 'yes',
		Method = rep(rna.fusions$Tool, times = 2),
		Frame = rep(rna.fusions$Frame, times = 2),
		Fusion_Status = rep(rna.fusions$Status, times = 2),
#		SV_Type = rep(rna.fusions$event_type, times = 2),
#		Seq_Type = rep(rna.fusions$Seq_Type, times = 2),
		stringsAsFactors = FALSE
		);

	if (is.null(cbio.data)) { cbio.data <- rna.cbio; }
	else {
		unique.rna <- unique(rna.fusions[,c('Sample','Fusion','Tool','Status','Frame','Support')]);
		for (i in 1:nrow(unique.rna)) {
			fusion <- unique.rna[i,]$Fusion;
			smp <- unique.rna[i,]$Sample
			dna.idx <- which(cbio.data$Fusion == fusion & cbio.data$Tumor_Sample_Barcode == smp);
			rna.idx <- which(rna.cbio$Fusion == fusion & rna.cbio$Tumor_Sample_Barcode == smp);
			if (length(dna.idx) > 0) {
	#			stop();
				cbio.data[dna.idx,]$RNA_support <- 'yes';
				cbio.data[dna.idx,]$Method <- paste0(
					unique(c(
						unlist(strsplit(cbio.data[dna.idx,]$Method,',')),
						unlist(strsplit(rna.cbio[rna.idx,]$Method,','))
						)),
					collapse = ','
					);
				}
			else {
				cbio.data <- rbind(cbio.data, rna.cbio[rna.idx,]);
				}
			}
		}
	}

to.write <- unique(cbio.data[order(cbio.data$Fusion, cbio.data$Tumor_Sample_Barcode),]);
to.write[which(to.write$Frame == 'unknown'),]$Frame <- NA;

write.table(
	to.write,
	file = generate.filename(arguments$project, 'svs_for_cbioportal','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

### SAVE SESSION INFO ##############################################################################
save.session.profile(generate.filename('CollectMAVIS','SessionProfile','txt'));
