### collect_rsem_output.R ##########################################################################
# Finds and combines output generated by RSEM.

### FUNCTIONS ######################################################################################
# function to generate a standardized filename
generate.filename <- function(project.stem, file.core, extension, include.date = TRUE) {

	# build up the filename
	file.name <- paste(project.stem, file.core, sep = '_');
	file.name <- paste(file.name, extension, sep = '.');

	if (include.date) {
		file.name <- paste(Sys.Date(), file.name, sep = '_');
		}

	return(file.name);
	}

# function to write session profile to file
save.session.profile <- function(file.name) {

	# open the file
	sink(file = file.name, split = FALSE);

	# write memory usage to file
	cat('### MEMORY USAGE ###############################################################');
	print(proc.time());

	# write sessionInfo to file
	cat("\n### SESSION INFO ###############################################################");
	print(sessionInfo());

	# close the file
	sink();

	}

# add entrez gene ID based on symbol
annotate.by.symbol <- function(symbols) {
	mapIds(org.Hs.eg.db,
		keys = symbols,
		keytype = 'SYMBOL',
		column = 'ENTREZID',
		multiVals = 'first'
		);
	}

### PREPARE SESSION ################################################################################
# import libraries
library(org.Hs.eg.db);
library(argparse);
library(plyr);

# import command line arguments
parser <- ArgumentParser();

parser$add_argument('-d', '--directory', type = 'character', help = 'path to data directory');
parser$add_argument('-p', '--project', type = 'character', help = 'project name');
parser$add_argument('-g', '--gtf', type = 'character', help = 'annotation gtf (refGene)',
        default = '/cluster/projects/pughlab/references/gencode/GRCh38/gencode.v31.GRCh38.genes.gtf');

arguments <- parser$parse_args();

# what's the date?
date <- Sys.Date();

setwd(arguments$directory);

### FORMAT ANNOTATION
refGene <- read.delim(arguments$gtf, header = FALSE, comment.char = '#');
refGene <- refGene[which(refGene$V3 == 'gene'),c(1,4,5,9)];
colnames(refGene) <- c('Chromosome','Start','End','INFO');

refGene$GeneID <- sapply(
        refGene$INFO,
        function(i) {
                parts <- unlist(strsplit(as.character(i), ';'));
                gene_name <- unlist(strsplit(parts[grepl('gene_id', parts)], ' '));
                gene_id  <- gene_name[length(gene_name)];
                return(gene_id);
                }
        );

refGene$Symbol <- sapply(
        refGene$INFO,
        function(i) {
                parts <- unlist(strsplit(as.character(i), ';'));
                gene_name <- unlist(strsplit(parts[grepl('gene_name', parts)], ' '));
                gene_symbol <- gene_name[length(gene_name)];
                return(gene_symbol);
                }
        );

refGene$Type <- sapply(
	refGene$INFO,
	function(i) {
		parts <- unlist(strsplit(as.character(i), ';'));
		gene_type <- unlist(strsplit(parts[grepl('gene_type', parts)], ' '));
		rna_type <- gene_type[length(gene_type)];
		return(rna_type);
		}
	);

refGene$Chromosome <- factor(refGene$Chromosome, levels = paste0('chr', c(1:22,'X','Y','M')));
refGene <- refGene[order(refGene$Chromosome, refGene$Start, refGene$End),];
refGene$GeneID <- factor(refGene$GeneID, levels = as.character(refGene$GeneID));

refGene$Entrez_Gene_ID <- annotate.by.symbol(refGene$Symbol);

### MAIN ###########################################################################################
# find results files
genes.files <- list.files(pattern = 'genes.results', recursive = TRUE);
isoforms.files <- list.files(pattern = 'isoforms.results', recursive = TRUE);

# read them in
genes.list <- list();
isoforms.list <- list();

for (file in genes.files) {
	# extract sample ID
	smp <- unlist(strsplit(file, '\\/'))[2];
	# read in data
	tmp <- read.delim(file);
	# store data in list
	genes.list[[smp]] <- tmp[,c('gene_id', 'transcript_id.s.', 'expected_count', 'TPM', 'FPKM')];
	colnames(genes.list[[smp]]) <- c('gene_id', 'transcript_id', paste0(smp, c('.expected_count', '.TPM','.FPKM')));
	}

for (file in isoforms.files) {
	# extract sample ID
	smp <- unlist(strsplit(file, '\\/'))[2];
	# read in data
	tmp <- read.delim(file);
	# store
	isoforms.list[[smp]] <- tmp[,c('transcript_id', 'gene_id', 'expected_count', 'TPM', 'FPKM')];
	colnames(isoforms.list[[smp]]) <- c('transcript_id', 'gene_id', paste0(smp, c('.expected_count','.TPM','.FPKM')));
	}

# reshape/format data
genes.formatted <- list();
isoforms.formatted <- list();

genes.tmp <- join_all(
	genes.list,
	by = c('gene_id','transcript_id'),
	type = 'full',
	match = 'first'
	);

isoforms.tmp <- join_all(
	isoforms.list,
	by = c('transcript_id','gene_id'),
	type = 'full',
	match = 'first'
	);

# for gene data, clean up annotations
genes.tmp <- merge(
	refGene[,c('GeneID','Symbol','Entrez_Gene_ID','Type')],
	genes.tmp,
	by.x = 'GeneID',
	by.y = 'gene_id'
	);

genes.tmp$GeneID <- factor(genes.tmp$GeneID, levels = levels(refGene$GeneID));
genes.tmp <- genes.tmp[order(genes.tmp$GeneID),];

genes.formatted$fpkm <- genes.tmp[,c(1:3, grep('FPKM', colnames(genes.tmp)))];
genes.formatted$tpm  <- genes.tmp[,c(1:3, grep('TPM', colnames(genes.tmp)))];
genes.formatted$expected_count  <- genes.tmp[,c(1:3, grep('expected_count', colnames(genes.tmp)))];

colnames(genes.formatted$fpkm) <- gsub('.FPKM', '', colnames(genes.formatted$fpkm));
colnames(genes.formatted$tpm)  <- gsub('.TPM', '', colnames(genes.formatted$tpm));
colnames(genes.formatted$expected_count)  <- gsub('.expected_count', '', colnames(genes.formatted$expected_count));

isoforms.formatted$fpkm <- isoforms.tmp[,c(1:3, grep('FPKM', colnames(isoforms.tmp)))];
isoforms.formatted$tpm  <- isoforms.tmp[,c(1:3, grep('TPM', colnames(isoforms.tmp)))];
isoforms.formatted$expected_count  <- isoforms.tmp[,c(1:3, grep('expected_count', colnames(isoforms.tmp)))];

colnames(isoforms.formatted$fpkm) <- gsub('.FPKM', '', colnames(isoforms.formatted$fpkm));
colnames(isoforms.formatted$tpm)  <- gsub('.TPM', '', colnames(isoforms.formatted$tpm));
colnames(isoforms.formatted$expected_count)  <- gsub('.expected_count', '', colnames(isoforms.formatted$expected_count));

# save combined/formatted data to file
write.table(
	genes.formatted$tpm,
	file = generate.filename(arguments$project, 'gene_expression_TPM','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

write.table(
	genes.formatted$expected_count,
	file = generate.filename(arguments$project, 'rsem_expected_counts','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

save(
	genes.formatted,
	isoforms.formatted,
	file = generate.filename(arguments$project, 'rsem_expression_results','RData')
	);

# prep or cBioportal
cbio <- genes.formatted$tpm;

all.zero <- apply(cbio[,4:ncol(cbio)],1,sum);
cbio <- cbio[which(all.zero > 0),];

cbio <- cbio[!grepl("PAR_Y", cbio$GeneID),];

# remove 'misc_RNA' without entrez gene ID
remove.ids <- unique(refGene[which(refGene$Type == 'misc_RNA'),]$GeneID);
cbio <- cbio[-which(cbio$GeneID %in% remove.ids & is.na(cbio$Entrez_Gene_ID)),];

if (any(duplicated(cbio$Symbol))) {

	dup.genes <- unique(cbio[duplicated(cbio$Symbol),]$Symbol);

	if ('ZNF883' %in% dup.genes) {
		cbio <- cbio[-which(cbio$Symbol == 'ZNF883' & !grepl('ENSG00000228623', cbio$GeneID)),];
		}
	if ('GGT1' %in% dup.genes) {
		cbio <- cbio[-which(cbio$Symbol == 'GGT1' & !grepl('ENSG00000100031', cbio$GeneID)),];
		}
	if ('LINC01238' %in% dup.genes) {
		cbio <- cbio[-which(cbio$Symbol == 'LINC01238' & !grepl('ENSG00000237940', cbio$GeneID)),];
		}
	if ('TMSB15B' %in% dup.genes) {
		cbio <- cbio[-which(cbio$Symbol == 'TMSB15B' & !grepl('ENSG00000269226', cbio$GeneID)),];
		}
	if ('C2orf27A' %in% dup.genes) {
		cbio <- cbio[-which(cbio$Symbol == 'C2orf27A' & !grepl('ENSG00000287151', cbio$GeneID)),];
		}
	if ('ARMCX5-GPRASP2' %in% dup.genes) {
		cbio <- cbio[-which(cbio$Symbol == 'ARMCX5-GPRASP2' & !grepl('ENSG00000286237', cbio$GeneID)),];
		}
	if ('DIABLO' %in% dup.genes) {
		cbio <- cbio[-which(cbio$Symbol == 'DIABLO' & !grepl('ENSG00000184047', cbio$GeneID)),];
		}
	if ('TBCE' %in% dup.genes) {
		cbio <- cbio[-which(cbio$Symbol == 'TBCE' & !grepl('ENSG00000284770', cbio$GeneID)),];
		}
	if ('GOLGA8M' %in% dup.genes) {
		cbio <- cbio[-which(cbio$Symbol == 'GOLGA8M' & !grepl('ENSG00000188626', cbio$GeneID)),];
		}
	if ('HSPA14' %in% dup.genes) {
		cbio <- cbio[-which(cbio$Symbol == 'HSPA14' & !grepl('ENSG00000187522', cbio$GeneID)),];
		}

	dup.genes <- unique(cbio[duplicated(cbio$Symbol),]$Symbol);

	cbio <- cbio[-which(cbio$Symbol %in% dup.genes & is.na(cbio$Entrez_Gene_ID)),];

	dup.genes <- unique(cbio[duplicated(cbio$Symbol),]$Symbol);
	if (length(dup.genes) > 0) {
		cbio$var <- apply(cbio[,4:ncol(cbio)],1,var);
		cbio <- cbio[order(cbio$Symbol, -cbio$var),];
		cbio <- cbio[!duplicated(cbio$Symbol),];
		}
	}

# format for TIMER
timer.data <- cbio[,c(2,4:ncol(cbio))]; 
rownames(timer.data) <- cbio$Symbol;
timer.data <- timer.data[,-1];

write.table(
	timer.data,
	file = generate.filename(arguments$project, 'mRNA_expression_TPM_for_TIMER','tsv'),
	row.names = TRUE,
	col.names = NA,
	quote = FALSE,
	sep = '\t'
	);

# set 0's to NA and log2 transform
cbio[which(cbio == 0, arr.ind = TRUE)] <- NA;
cbio <- cbind(
	cbio[,2:3],
	log2(cbio[,4:ncol(cbio)])
	);
colnames(cbio)[1] <- 'Hugo_Symbol';
colnames(cbio)[2] <- 'Entrez_Gene_Id';

write.table(
	cbio,
	file = generate.filename(arguments$project, 'mRNA_expression_TPM_for_cbioportal','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

# and do a default scaling
zscores <- data.frame(
	cbio[,1:2],
	t(apply(t(cbio[,3:ncol(cbio)]),2,scale))
	);
colnames(zscores) <- colnames(cbio);

all.na <- apply(zscores[,3:ncol(zscores)],1,function(i) { all(is.na(i)) } );

write.table(
	zscores[which(all.na == FALSE),],
	file = generate.filename(arguments$project, 'mRNA_TPM_zscores_for_cbioportal','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

### SAVE SESSION INFO ##############################################################################
save.session.profile(generate.filename('CollectRSEM','SessionProfile','txt'));
