### collect_mavis_output.R #########################################################################
# Finds and combines output generated by MAVIS.

### FUNCTIONS ######################################################################################
# function to generate a standardized filename
generate.filename <- function(project.stem, file.core, extension, include.date = TRUE) {

	# build up the filename
	file.name <- paste(project.stem, file.core, sep = '_');
	file.name <- paste(file.name, extension, sep = '.');

	if (include.date) {
		file.name <- paste(Sys.Date(), file.name, sep = '_');
		}

	return(file.name);
	}

# function to write session profile to file
save.session.profile <- function(file.name) {

	# open the file
	sink(file = file.name, split = FALSE);

	# write memory usage to file
	cat('### MEMORY USAGE ###############################################################');
	print(proc.time());

	# write sessionInfo to file
	cat("\n### SESSION INFO ###############################################################");
	print(sessionInfo());

	# close the file
	sink();

	}

# function to annotate entrez gene ids by symbol
annotate.by.symbol <- function(hugo_symbols) {
	mapIds(org.Hs.eg.db,
		keys = hugo_symbols,
		column = "ENTREZID",
		keytype = "SYMBOL",
		multiVals = "first"
		);
	}

### PREPARE SESSION ################################################################################
# import libraries
library(GenomicRanges);
library(argparse);
library(plyr);
library(AnnotationDbi);
library(org.Hs.eg.db);

# import command line arguments
parser <- ArgumentParser();

parser$add_argument('-d', '--directory', type = 'character', help = 'path to data directory');
parser$add_argument('-p', '--project', type = 'character', help = 'project name');
parser$add_argument('-t', '--targets', type = 'character', help = 'path to target regions', default = NULL);
parser$add_argument('-g', '--genome', type = 'character', help = 'genome build', default = 'GRCh38');
parser$add_argument('-c', '--center', type = 'character', help = 'data center (for cbioportal annotation)', default = 'PMCC');
parser$add_argument('-f', '--find_drawings', type = 'logical', help = 'find and extract key drawings?', default = FALSE);

arguments <- parser$parse_args();

# what's the date?
date <- Sys.Date();

setwd(arguments$directory);

### COLLECT KEY DRAWINGS ###########################################################################
if (arguments$find_drawings) {

	# move to input directory (should be patient directory)
	input.dir <- getwd();
	output.dir <- paste0(input.dir, '/key_drawings');

	# create new folder
	dir.create(output.dir);

	# find all svg files
	all.drawings <- list.files(path = input.dir, pattern = '.svg$', recursive = TRUE, full.names = TRUE);

	print(paste('Found', length(all.drawings), 'svg files.'));

	# filter drawings (remove any intergenic ones)
	key.drawings <- all.drawings[!grepl('-NA_|_NA.svg', all.drawings)];

	print(paste('  Copying', length(key.drawings), 'gene-gene .svg files to', output.dir, '...'));

	setwd(output.dir);

	for (i in key.drawings) {
		cp.cmd <- paste0('cp ', i, ' .');
		system(cp.cmd);
		}

	} else {

### MAIN ###########################################################################################
	# find results files
	mavis.files <- list.files(pattern = 'mavis_summary_all', recursive = TRUE);

	# read them in
	sv.list <- list();

	for (file in mavis.files) {
		# extract sample ID
		smp <- unlist(strsplit(file, '\\/'))[1];
		# read in and store data in list
		sv.list[[smp]] <- read.delim(file, as.is = TRUE);
		}

	# reshape/format data
	sv.data.full <- join_all(
		sv.list,
		type = 'full',
		match = 'first'
		);
	colnames(sv.data.full)[which(colnames(sv.data.full) == 'X.tracking_id')] <- 'tracking_id';

	key.fields <- c('event_type','gene1_aliases','gene2_aliases','transcript1','transcript2','gene1_direction','gene2_direction','exon_last_5prime','exon_first_3prime','fusion_splicing_pattern','break1_chromosome','break1_position_start','break1_position_end','break2_chromosome','break2_position_start','break2_position_end','break1_split_reads','break2_split_reads','spanning_reads','flanking_pairs','linking_split_reads','library','tracking_id','tools','protocol');

	smp.fields <- sort(colnames(sv.data.full)[grep('_genome|_transcriptome',colnames(sv.data.full))]);
	smp.fields.names <- sapply(smp.fields, function(i) { unlist(strsplit(i,'_'))[1] } );
	normal.smps <- gsub('\\.','-',smp.fields.names[grepl('normal', smp.fields)]);
	tumour.smps <- gsub('\\.','-',smp.fields.names[grepl('diseased_genome', smp.fields)]);
	rna.smps <- gsub('\\.','-',smp.fields.names[grepl('transcriptome', smp.fields)]);

	sv.data <- sv.data.full[,c(key.fields, smp.fields)];

	# fill in missing tools (custom conversion scripts)
	new.tools <- apply(
		sv.data[,c('tracking_id','tools')],
		1,
		function(i) {

			ids <- unlist(strsplit(as.character(i[1]),';'));
			tools <- setdiff(unlist(strsplit(as.character(i[2]),';')),'');

			if (any(grepl('pindel', ids))) { tools <- c(tools,'pindel'); }
			if (any(grepl('svict', ids))) { tools <- c(tools,'svict'); }
			if (any(grepl('novobreak', ids))) { tools <- c(tools,'novobreak'); }
			if (any(grepl('fusioncatcher', ids))) { tools <- c(tools,'fusioncatcher'); }

			toolset <- paste(tools, collapse = ';');
			if (length(tools) > 0) { 
				return(toolset);
				} else {
				return('');
				}
			}
		);

	sv.data$tools <- new.tools;
	sv.data$Fusion <- paste0(sv.data$gene1_aliases,'--',sv.data$gene2_aliases);

	# save combined/formatted data to file
	write.table(
		sv.data,
		file = generate.filename(arguments$project, 'mavis_output','tsv'),
		row.names = FALSE,
		col.names = TRUE,
		sep = '\t'
		);

	# if target regions were provided
	if (!is.null(arguments$targets)) {

		# get target regions
		target_bed <- read.delim(arguments$targets, header = FALSE, comment.char = '#');
		colnames(target_bed)[1:3] <- c('Chromosome','Start','End');

		# create genomic ranges object for target regions
		target.gr <- makeGRangesFromDataFrame(target_bed, starts.in.df.are.0based = TRUE);

		# create genomic ranges object for each breakpoint
		first_bp <- data.frame(
			Chromosome = paste0('chr',sv.data$break1_chromosome),
			Start = sv.data$break1_position_start,
			End = sv.data$break1_position_end
			);

		second_bp <- data.frame(
			Chromosome = paste0('chr',sv.data$break2_chromosome),
			Start = sv.data$break2_position_start,
			End = sv.data$break2_position_end
			);

		bp1.gr <- makeGRangesFromDataFrame(first_bp, starts.in.df.are.0based = FALSE);
		bp2.gr <- makeGRangesFromDataFrame(second_bp, starts.in.df.are.0based = FALSE);

		# find overlaps
		overlaps.p1 <- as.data.frame(findOverlaps(bp1.gr, target.gr));
		overlaps.p2 <- as.data.frame(findOverlaps(bp2.gr, target.gr));

		overlap.data <- merge(
			overlaps.p1,
			overlaps.p2,
			by = 'queryHits',
			suffixes = c('.1','.2'),
			all = TRUE
			);

		# only keep entries for which both breakpoints are within target regions
		to.remove <- which(is.na(overlap.data$subjectHits.1) | is.na(overlap.data$subjectHits.2));
		keep.idx <- unique(overlap.data[-to.remove,]$queryHits);

		# filter initial input
		sv.data.filtered <- sv.data[keep.idx,];

		# save filtered data
		write.table(
			sv.data.filtered,
			file = generate.filename(arguments$project, 'mavis_output_filtered','tsv'),
			row.names = FALSE,
			col.names = TRUE,
			sep = '\t'
			);

		} else { 
		sv.data.filtered <- sv.data;
		}

	### CBIOPORTAL ######################
	# format SV data for cbioportal (NEW!)
	tmp <- sv.data.filtered[which(sv.data.filtered$gene1_aliases != 'None' &
		sv.data.filtered$gene2_aliases != 'None'),];

	# sort by evidence
	tmp$N.tools <- sapply(tmp$tools, function(i) { length(unlist(strsplit(i,';'))) } );
	tmp$Evidence <- apply(tmp[,grep('reads|pairs', colnames(tmp))],1, function(i) { 
		sum(sapply(i[which(i != 'None')],function(y) { max(as.numeric(unlist(strsplit(y,';')))) } ))
		} );

	tmp <- tmp[order(tmp$library, tmp$tracking_id, -tmp$N.tools, -tmp$Evidence),];

	# is this a tumour or normal sample?
	tmp$Status <- 'somatic';
	if (length(normal.smps) > 0) {
		tmp[which(tmp$library %in% normal.smps),]$Status <- 'germline';
		}

	# is this an inframe or frameshift variant?
	tmp$Frame <- 'unknown';
	tmp[which(tmp$fusion_splicing_pattern == 'normal'),]$Frame <- 'inframe';
	tmp[which(!tmp$fusion_splicing_pattern %in% c('normal','None')),]$Frame <- 'frameshift';

	# trim to unique events (most evidence wins)
	dna <- tmp[which(tmp$protocol == 'genome'),];
	rna <- if (any(grepl('transcriptome', tmp$protocol))) {
		tmp[which(tmp$protocol == 'transcriptome'),];
		} else { NULL; }

	somatic.svs <- dna[which(dna$Status == 'somatic'),];
	somatic.svs <- somatic.svs[!duplicated(somatic.svs[,c('library','tracking_id')]),];
	germline.svs <- dna[which(dna$Status == 'germline'),];
	germline.svs <- germline.svs[!duplicated(germline.svs[,c('library','tracking_id')]),];

	# is this truely somatic, or observed in matched normal?
	germ.idx <- apply(somatic.svs[,smp.fields],1,function(i) { any(i == 'germline', na.rm = TRUE) } );
	somatic.svs[germ.idx,]$Status <- 'germline';

	# add in RNA support if present
	if (!is.null(rna) & (length(rna.smps) > 0)) {

		for (i in 1:nrow(rna)) {
			fusion <- as.character(rna[i,]$Fusion);
			type <- as.character(rna[i,]$event_type);
			smp <- sub('-rna','',rna[i,]$library);
			exon1 <- rna[i,]$exon_last_5prime;
			exon2 <- rna[i,]$exon_first_3prime;

			dna.idx <- which(somatic.svs$Fusion == fusion &
				somatic.svs$event_type == type &
				somatic.svs$library == smp &
				somatic.svs$exon_last_5prime == exon1 &
				somatic.svs$exon_first_3prime == exon2
				);

			if (length(dna.idx) == 1) {
				somatic.svs[dna.idx,]$tools <- paste(
					unique(c(unlist(strsplit(somatic.svs[dna.idx,]$tools,';|,')),
						unlist(strsplit(rna[i,]$tools,';|,')))),
					collapse = ','
					);
				} else if (length(dna.idx) == 0) {
				somatic.svs <- rbind(somatic.svs, rna[i,]);
				} else if (length(dna.idx) > 1) {
				stop();
				}
			}
		}

	somatic.svs <- somatic.svs[!duplicated(somatic.svs[,c('library','tracking_id')]),];

	# format for cBioportal
	cbio.svs <- data.frame(
		Sample_ID = gsub('-rna|-wgs|-wxs','',somatic.svs$library),
		Site1_Hugo_Symbol = somatic.svs$gene1_aliases,
		Site1_Entrez_Gene_Id = annotate.by.symbol(somatic.svs$gene1_aliases),
		Site1_Ensembl_Transcript_Id = somatic.svs$transcript1,
		Site1_Exon = somatic.svs$exon_last_5prime,
		Site1_Chromosome = somatic.svs$break1_chromosome,
		Site1_Position = somatic.svs$break1_position_start,
		Site1_Description = NA,
		Site2_Hugo_Symbol = somatic.svs$gene2_aliases,
		Site2_Entrez_Gene_Id = annotate.by.symbol(somatic.svs$gene2_aliases),
		Site2_Ensembl_Transcript_Id = somatic.svs$transcript2,
		Site2_Exon = somatic.svs$exon_first_3prime,
		Site2_Chromosome = somatic.svs$break2_chromosome,
		Site2_Position = somatic.svs$break2_position_start,
		Site2_Description = NA,
		Site2_Effect_On_Frame = toupper(somatic.svs$Frame),
		NCBI_Build = rep(arguments$genome,nrow(somatic.svs)),
		DNA_Support = rep('no', nrow(somatic.svs)),
		RNA_Support = rep('no', nrow(somatic.svs)), 
		Normal_Read_Count = NA,
		Tumor_Read_Count = NA,
		Normal_Variant_count = NA,
		Tumor_Variant_Count = NA,
		Normal_Paired_End_Read_Count = NA,
		Tumor_Paired_End_Read_Count = NA,
		Normal_Split_Read_Count = NA,
		Tumor_Split_Read_Count = NA,
		Annotation = NA,
		Breakpoint_Type = NA,
		Connection_Type = paste0(somatic.svs$gene1_direction,'to',somatic.svs$gene2_direction),
		Event_Info = NA, 
		Class = toupper(somatic.svs$event_type),
		Length = NA,
		Comments = somatic.svs$Status,
		Fusion = paste0(somatic.svs$gene1_aliases, '--', somatic.svs$gene2_aliases)
		);

	# fill in support types
	cbio.svs$DNA_Support <- as.character(cbio.svs$DNA_Support);
	cbio.svs$RNA_Support <- as.character(cbio.svs$RNA_Support);
	dna.support <- grepl('manta|delly|pindel|novobreak|svict', somatic.svs$tools);
	rna.support <- grepl('starfusion|fusioncatcher', somatic.svs$tools);
	cbio.svs[dna.support,]$DNA_Support <- 'yes';
	cbio.svs[rna.support,]$RNA_Support <- 'yes';

	# extract breakpoint type (precise/imprecise)
#	cbio.svs$Breakpoint_Type <- 'IMPRECISE';
#	precise.idx <- which(somatic.svs$break1_position_start == somatic.svs$break1_position_end &
#		somatic.svs$break2_position_start == somatic.svs$break2_position_end);
#	cbio.svs[precise.idx,]$Breakpoint_Type <- 'PRECISE';

	# clean up Frame status
	cbio.svs$Site2_Effect_On_Frame <- factor(cbio.svs$Site2_Effect_On_Frame,
		levels = c('INFRAME','FRAMESHIFT')
		);

	# calculate intrachromosomal SV lengths
	cbio.svs$Site1_Chromosome <- as.character(cbio.svs$Site1_Chromosome);
	cbio.svs$Site2_Chromosome <- as.character(cbio.svs$Site2_Chromosome);
	
	intra.chrom <- which(cbio.svs$Site1_Chromosome == cbio.svs$Site2_Chromosome);
	cbio.svs[intra.chrom,]$Length <- abs(
		cbio.svs[intra.chrom,]$Site1_Position - cbio.svs[intra.chrom,]$Site2_Position);

	# save to file
	write.table(
		cbio.svs,
		file = generate.filename(arguments$project, 'sv_data_for_cbioportal','tsv'),
		row.names = FALSE,
		col.names = TRUE,
		quote = FALSE,
		sep = '\t'
		);

	gc();

	### OLD VERSION ###
	# format for cbioportal (fusion_data.txt > DEPRECATED in cbioportal)
	fusion.data <- data.frame(
		Hugo_Symbol = c(
			as.character(cbio.svs$Site1_Hugo_Symbol),
			as.character(cbio.svs$Site2_Hugo_Symbol)
			),
		Entrez_Gene_Id = c(cbio.svs$Site1_Entrez_Gene_Id, cbio.svs$Site2_Entrez_Gene_Id),
		Center = rep(arguments$center, nrow(cbio.svs)*2),
		Tumor_Sample_Barcode = rep(cbio.svs$Sample_ID, times = 2),
		Fusion = rep(cbio.svs$Fusion, times = 2),
		DNA_Support = rep(cbio.svs$DNA_Support, times = 2),
		RNA_Support = rep(cbio.svs$RNA_Support, times = 2),
		Method = rep(gsub(';', ',', somatic.svs$tools), times = 2),
		Frame = tolower(rep(cbio.svs$Site2_Effect_On_Frame, times = 2)),
		Fusion_Status = rep(cbio.svs$Comments, times = 2)
		);

	# save to file
	write.table(
		fusion.data,
		file = generate.filename(arguments$project, 'fusion_data_for_cbioportal','tsv'),
		row.names = FALSE,
		col.names = TRUE,
		quote = FALSE,
		sep = '\t'
		);

	### SAVE SESSION INFO ##############################################################################
	save.session.profile(generate.filename('CollectMAVIS','SessionProfile','txt'));
	}
