### collect_gatk_gcnv_output.R #####################################################################
# Finds and combines output generated by GATK:gCNV

### FUNCTIONS ######################################################################################
# function to generate a standardized filename
generate.filename <- function(project.stem, file.core, extension, include.date = TRUE) {

	# build up the filename
	file.name <- paste(project.stem, file.core, sep = '_');
	file.name <- paste(file.name, extension, sep = '.');

	if (include.date) {
		file.name <- paste(Sys.Date(), file.name, sep = '_');
		}

	return(file.name);
	}

# function to write session profile to file
save.session.profile <- function(file.name) {

	# open the file
	sink(file = file.name, split = FALSE);

	# write memory usage to file
	cat('### MEMORY USAGE ###############################################################');
	print(proc.time());

	# write sessionInfo to file
	cat("\n### SESSION INFO ###############################################################");
	print(sessionInfo());

	# close the file
	sink();

	}

### PREPARE SESSION ################################################################################
# import libraries
library(plyr);
library(org.Hs.eg.db);
library(argparse);

# import command line arguments
parser <- ArgumentParser();

parser$add_argument('-d', '--directory', type = 'character', help = 'path to data directory');
parser$add_argument('-p', '--project', type = 'character', help = 'project name');
parser$add_argument('-r', '--ref_type', type = 'character', help = 'reference type', default = 'hg38');
parser$add_argument('-t', '--targets', type = 'character', help = 'target intervals');

arguments <- parser$parse_args();

# what's the date?
date <- Sys.Date();

setwd(arguments$directory);

### FORMAT ANNOTATION ##############################################################################
if (arguments$ref_type %in% c('hg38','GRCh38')) {
	library(TxDb.Hsapiens.UCSC.hg38.knownGene);
	txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene;
	} else if (arguments$ref_type %in% c('hg19','GRCh37')) {
	library(TxDb.Hsapiens.UCSC.hg19.knownGene);
	txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene;
	}
 
# organize all transcripts by GeneID
txdb_table <- transcriptsBy(txdb, by = 'gene');

# extract GeneIDs
tx_ids <- names(txdb_table);

# extract 
transcript_table <- select(
	TxDb.Hsapiens.UCSC.hg38.knownGene,
	keys = tx_ids,
	columns = c('GENEID','TXNAME','TXCHROM','TXSTART','TXEND'),
	keytype = 'GENEID'
	);
colnames(transcript_table)[1] <- 'ENTREZID';

# pull in extra annotations
transcript_table$SYMBOL <- mapIds(
	org.Hs.eg.db,
	keys = transcript_table$TXNAME,
	keytype = 'UCSCKG',
	column = 'SYMBOL'
	);

transcript_table$ENSEMBL <- mapIds(
	org.Hs.eg.db,
	keys = transcript_table$TXNAME,
	keytype = 'UCSCKG',
	column = 'ENSEMBL'
	);

transcript_table$MAP <- mapIds(
	org.Hs.eg.db,
	keys = transcript_table$TXNAME,
	keytype = 'UCSCKG',
	column = 'MAP'
	);

transcript_table$GENETYPE <- mapIds(
	org.Hs.eg.db,
	keys = transcript_table$TXNAME,
	keytype = 'UCSCKG',
	column = 'GENETYPE'
	);

# remove unmatching cases
transcript_table$MAPCHROM <- sapply(
	transcript_table$MAP,
	function(i) { paste0('chr', unlist(strsplit(i, 'p|q'))[1]) }
	);

transcript_table <- transcript_table[which(transcript_table$TXCHROM == transcript_table$MAPCHROM),];

# squish it to 1 entry per gene
gene.annotation <- merge(
	aggregate(TXSTART ~ ENTREZID + SYMBOL + ENSEMBL + GENETYPE + TXCHROM, transcript_table, min),
	aggregate(TXEND ~ ENTREZID + SYMBOL + ENSEMBL + GENETYPE + TXCHROM, transcript_table, max),
	);
colnames(gene.annotation)[5:7] <- c('Chromosome','Start','End');
gene.annotation$Chromosome <- factor(gene.annotation$Chromosome, levels = paste0('chr',c(1:22,'X','Y')));
gene.annotation <- gene.annotation[order(gene.annotation$Chromosome, gene.annotation$Start),];

gene.gr <- GRanges(gene.annotation);

# if provided, filter to target intervals
gene.annotation$Target <- NA;
if (!is.null(arguments$targets)) {

	target.intervals <- read.delim(arguments$targets, header = F, comment.char = '#');
	colnames(target.intervals)[1:3] <- c('Chromosome','Start','End');

	target.gr <- GRanges(target.intervals);

	overlaps <- as.data.frame(findOverlaps(gene.gr, target.gr));

	gene.annotation$Target <- 0;
	gene.annotation[unique(overlaps$queryHits),]$Target <- 1;
	}

### MAIN ###########################################################################################
# find results files
ratio.files <- list.files(pattern = 'denoised_copy_ratios.tsv', recursive = TRUE);
seg.files <- list.files(pattern = 'genotyped_segments.vcf$', recursive = TRUE);

# read them in
ratio.list <- list();

for (file in ratio.files) {
	# extract sample ID
	smp <- unlist(strsplit(basename(file), '\\_'))[1];
	# store data in list
	tmp <- read.delim(file, as.is = TRUE, comment.char = '@');
	colnames(tmp)[4] <- smp;
	ratio.list[[smp]] <- tmp;
	rm(tmp);
	}

ratio.data <- join_all(ratio.list);

colnames(ratio.data)[1:3] <- c('chrom','start','end');
seg.gr <- GRanges(ratio.data);
y <- mergeByOverlaps(gene.gr, seg.gr, minoverlap = 500, select = 'all');
tmp <- data.frame(y[['gene.gr']], y[['seg.gr']]);
colnames(tmp)[1:5] <- paste0('gene.',colnames(tmp)[1:5]);
colnames(tmp)[10:14] <- paste0('seg.', gsub('\\.1','',colnames(tmp)[10:14]));

gene.by.patient <- aggregate(
	tmp[,grepl('INS',colnames(tmp))],
	by = list(
		SYMBOL = tmp$SYMBOL,
		ENTREZID = tmp$ENTREZID,
		ENSEMBL = tmp$ENSEMBL
		),
	FUN = mean
	);

ratio.data.annotated <- merge(
	gene.annotation,
	gene.by.patient,
	all.y = TRUE
	);

ratio.data.annotated$SYMBOL <- factor(
	ratio.data.annotated$SYMBOL,
	levels = as.character(gene.annotation$SYMBOL)
	);

write.table(
	ratio.data.annotated[order(ratio.data.annotated$SYMBOL),],
	file = generate.filename(arguments$project, 'gatk_gcnv_ratio_data','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

# read them in
cn.list <- list();

for (file in seg.files) {
	# extract sample ID
	smp <- unlist(strsplit(basename(file), '\\_'))[1];
	# store data in list
	tmp <- read.delim(file, as.is = TRUE, comment.char = '#', header = FALSE);
	colnames(tmp) <- c('chrom','start','id','ref','alt','qual','filter','info','format','sample');
	tmp$end <- as.numeric(gsub('END=','',tmp$info));
	tmp$QA <- apply(tmp[,c('format','sample')], 1, function(i) {
		idx <- which(unlist(strsplit(i[1],':')) == 'QA');
		quality <- unlist(strsplit(i[2],':'))[idx];
		return(as.numeric(quality));
		} );
	tmp <- tmp[which(tmp$QA >= 20),];
	tmp$CN <- apply(tmp[,c('format','sample')], 1, function(i) {
		idx <- which(unlist(strsplit(i[1],':')) == 'CN');
		cn <- unlist(strsplit(i[2],':'))[idx];
		return(cn);
		} );
	cn.list[[smp]] <- tmp[,c('chrom','start','end','CN')];
	cn.list[[smp]]$ID <- smp;
	rm(tmp);
	}

cn.data <- do.call(rbind, cn.list);
cn.data$CN <- as.numeric(cn.data$CN);

write.table(
	cn.data[,c('ID','chrom','start','end','CN')],
	file = generate.filename(arguments$project, 'gatk_gcnv_data','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

seg.gr <- GRanges(cn.data);
y <- mergeByOverlaps(gene.gr, seg.gr, minoverlap = 500, type = 'any', select = 'all', maxgap = -1);
tmp <- data.frame(y[['gene.gr']], y[['seg.gr']]);
colnames(tmp)[1:5] <- paste0('gene.',colnames(tmp)[1:5]);
colnames(tmp)[10:14] <- paste0('seg.', gsub('\\.1','',colnames(tmp)[10:14]));

pergene.data <- aggregate(
	tmp$CN,
	by = list(
		SYMBOL = tmp$SYMBOL,
		ENTREZID = tmp$ENTREZID,
		ENSEMBL = tmp$ENSEMBL,
		ID = tmp$ID
		),
	FUN = function(i) { max(i)-2 }
	);

gene.by.patient <- reshape(
	pergene.data,
	direction = 'wide',
	timevar = 'ID',
	idvar = c('SYMBOL','ENTREZID','ENSEMBL')
	);

cn.data.annotated <- merge(
	gene.annotation,
	gene.by.patient,
	all.y = TRUE
	);

cn.data.annotated$SYMBOL <- factor(
	cn.data.annotated$SYMBOL,
	levels = as.character(gene.annotation$SYMBOL)
	);

write.table(
	cn.data.annotated[order(cn.data.annotated$SYMBOL),],
	file = generate.filename(arguments$project, 'gatk_gcnv_data_annotated','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

# define a results table
cn.data$ENSEMBL <- NA;

for (smp in names(cn.list)) {

	# convert to GRanges
	seg.gr <- GRanges(cn.data);

	# find overlap with genes (any amount of overlap)
	overlapGenes <- as.data.frame(findOverlaps(seg.gr, gene.gr, minoverlap = 200));

	for (i in unique(overlapGenes$subjectHits)) {
		gene <- gene.annotation[i,]$ENSEMBL;
		idx <- unique(gene.seg.overlaps[which(gene.seg.overlaps$subjectHits == i),]$queryHits);
		cn.data[idx,]$ENSEMBL <- gene;
		}
	}

write.table(
	cn.data[,c('ID','ENSEMBL','chrom','start','end','CN')],
	file = generate.filename(arguments$project, 'gatk_gcnv_data_annotated','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

pga.data <- do.call(rbind, pga.values);
colnames(pga.data)[1] <- 'PGA';

# save data to file
write.table(
	pga.data,
	file = generate.filename(arguments$project, 'gatk_pga_estimates','tsv'),
	row.names = TRUE,
	col.names = TRUE,
	sep = '\t'
	);

write.table(
	gene.data.cn,
	file = generate.filename(arguments$project, 'gatk_cna_gene_matrix', 'tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

write.table(
	gene.data.ratio,
	file = generate.filename(arguments$project, 'gatk_ratio_gene_matrix', 'tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

### SAVE SESSION INFO ##############################################################################
save.session.profile(generate.filename('CollectCNAs','SessionProfile','txt'));
