### collect_panelCN_mops_output.R ##################################################################
# Finds and combines output generated by PanelCN.mops

### FUNCTIONS ######################################################################################
# function to generate a standardized filename
generate.filename <- function(project.stem, file.core, extension, include.date = TRUE) {

	# build up the filename
	file.name <- paste(project.stem, file.core, sep = '_');
	file.name <- paste(file.name, extension, sep = '.');

	if (include.date) {
		file.name <- paste(Sys.Date(), file.name, sep = '_');
		}

	return(file.name);
	}

# function to write session profile to file
save.session.profile <- function(file.name) {

	# open the file
	sink(file = file.name, split = FALSE);

	# write memory usage to file
	cat('### MEMORY USAGE ###############################################################');
	print(proc.time());

	# write sessionInfo to file
	cat("\n### SESSION INFO ###############################################################");
	print(sessionInfo());

	# close the file
	sink();

	}

# function to determine amplification status of gene
summarizeCNmops <- function(cn_mops_results, gain_threshold = 2.1) {

	if (any(cn_mops_results$Gene == "") | any(is.na(cn_mops_results$Gene))) {
		warning(paste0("Some targets had no associated gene name in the provided bed", 
			"file and where therefore removed from gene-level amplification",
			"calling analysis."));
		}

	cn_mops_results <- cn_mops_results[!is.na(cn_mops_results$Gene),];
	cn_mops_results <- subset(cn_mops_results, Gene != "" & lowQual != "lowQual");

	genes <- unique(cn_mops_results$Gene);
	samples <- unique(cn_mops_results$Sample);

	summ <- matrix(
		data = "NOT_AMP",
		nrow = length(genes), 
		ncol = length(samples), 
		dimnames = list(genes,samples)
		);

	for (gene in genes) {
		for (samp in samples) {

			tmp <- subset(cn_mops_results, Sample == samp & Gene == gene);
			cn.data <- as.data.frame(table(tmp$CN));
			cn.data$Val <- as.numeric(gsub("CN","",cn.data$Var1));
			Weighted_Average_CN <- sum(cn.data$Val*cn.data$Freq) / sum(cn.data$Freq);

			if (Weighted_Average_CN > gain_threshold) {
				summ[gene, samp] <- "AMP";
				}
			}
		}

	return(summ);
	}

### PREPARE SESSION ################################################################################
# import libraries
library(argparse);

# import command line arguments
parser <- ArgumentParser();

parser$add_argument('-d', '--directory', type = 'character', help = 'path to data directory');
parser$add_argument('-p', '--project', type = 'character', help = 'project name');

arguments <- parser$parse_args();

# what's the date?
date <- Sys.Date();

setwd(arguments$directory);

### MAIN ###########################################################################################
# find results files
cn.files <- list.files(pattern = 'panelcn.mops_results.tsv$', recursive = TRUE);

# read them in
if (any(grepl('cohort',cn.files))) {
	cohort.data <- read.delim(cn.files[grepl('cohort',cn.files)]);
	cn.files <- cn.files[!grepl('cohort',cn.files)];
	} else {
	cohort.data <- NULL;
	}

cn.data <- do.call(rbind,
	lapply(
		cn.files,
		read.delim
		)
	);

# save raw data to file
write.table(
	cn.data,
	file = generate.filename(arguments$project, 'panelCN.mops_output','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

# format data
cn.data$CN <- as.numeric(gsub('CN','',as.character(cn.data$CN)));
cn.data[which(cn.data$lowQual == 'lowQual'),]$CN <- NA;
colnames(cn.data)[2] <- 'Chromosome';

# reshape discrete CN calls
cn.calls <- reshape(
	cn.data[,c('Chromosome','Gene','Exon','Start','End','Sample','CN')],
	direction = 'wide',
	idvar = c('Chromosome','Gene','Exon','Start','End'),
	timevar = 'Sample'
	);
colnames(cn.calls) <- gsub('CN.','',colnames(cn.calls));

# reshape normalized read counts
read.counts <- reshape(
	cn.data[,c('Chromosome','Gene','Exon','Start','End','Sample','RC.norm')],
	direction = 'wide',
	idvar = c('Chromosome','Gene','Exon','Start','End'),
	timevar = 'Sample'
	);
colnames(read.counts) <- gsub('RC.norm.','',colnames(read.counts));

# calculate log2ratios
#cn.data$Ratio <- log2(cn.data$RC.norm / cn.data$medRC.norm);

#ratio.data <- reshape(
#	cn.data[,c('Chromosome','Gene','Exon','Start','End','Sample','Ratio')],
#	direction = 'wide',
#	idvar = c('Chromosome','Gene','Exon','Start','End'),
#	timevar = 'Sample'
#	);
#colnames(ratio.data) <- gsub('Ratio.','',colnames(ratio.data));

# save combined/formatted data to file
write.table(
	cn.calls,
	file = generate.filename(arguments$project, 'panelCN.mops_cna_matrix','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

write.table(
	read.counts,
	file = generate.filename(arguments$project, 'panelCN.mops_normalized_readcounts','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

### SAVE SESSION INFO ##############################################################################
save.session.profile(generate.filename('CollectPanelCNmops','SessionProfile','txt'));
